---
title: "CDF Heatmap No Selenium for publication"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    df_print: paged
---

For this heatmap what we do is the following. From the bam files from each Selenoprotein we only keep reads longer than 25 pb and shorter than 45 pb and then we compute the coverage of each nutcleotide through all the transcript. With those values, we were able to calculate the cummulative fraction for each selenoprotein. Using cummulative values we calculated the differences between KO - WT for fixed bins down and up to SEc. How? We identified 20 % windows (100, 80, 60, 40, 20) from UTR start to Sec and from Sec to end of transcript. And on those windows with used the max cummulative fractions for each KO or WT and calculate fold change KO - WT. 

```{r, echo=F, message=F, warning=F}
#Load libraries
library(stringr)
library(plyr)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(grid)
library(decoupleR)
library(reshape2)
library(ComplexHeatmap)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE, echo = F)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```


```{r, echo=F, message=F, warning=F}
#Read the output of the coverage script'
#transcript name: MANE trancript name
#gene name: gene name
#transcript length: transcript length 
#sec_loc: selenocytine position at the transcript
heatmap_values_seleno <- function(transcript_name,gene_name,transcript_length,sec_loc){
cover_dir <- "/SSD/maa7095/RIBOseq/Process/Raw_data/RAW_FASTQ_RIBO_DATA/trimmed_out/non_coding_mapping/kallisto_output_kmer21/maped_bam_files"

  my_files <- list.files(paste0(cover_dir, "/",transcript_name), pattern="_no_se.*\\.cov")
  my_files2 <- str_extract(my_files, '.*(?=\\.cov)')
  #All the files in same format to read it correctly
  for (i in my_files2) {
    filename <- str_split(i, "_trimmed_")[[1]][2]
    wd <- paste0(cover_dir, "/", transcript_name, "/", i, ".cov")
    if (file.size(wd) > 0) {
      assign(filename, read.csv(wd, header=T, sep="\t"))
    } else {
      cat("File", wd, "is empty.\n")
    }
  }
  
  #Change the names of prob by each samples name.
  lst <-  mget(ls(pattern="RIBO"))
  lst1 <- Map(function(x,y) {names(x)[1] <- y; x},  lst, paste0(gene_name))
  lst1 <- Map(function(x,y) {names(x)[2] <- y; x},  lst1, "p5")
  lst1 <- Map(function(x,y) {names(x)[3] <- y; x},  lst1, "count")
  
  
  list2env(lst1, envir=.GlobalEnv)
  
  total_rows <- sum(unlist(lapply(lst1, function(df) dim(df)[1])))
  total_col <- sum(unlist(lapply(lst1, function(df) dim(df)[2]))) 
  
  add_missing_rows <- function(df, MISSING, transcript_name, gene_name) {
    df <- rbind(df,setNames(data.frame(paste0(gene_name), MISSING, 0), c(paste0(transcript_name), "p5", "count")))
    df[[""]] <- transcript_name
    df <- df[order(df$p5),]
    return(df)
  }
  
  lst1 <- lapply(lst1, function(df) {
    MISSING <- setdiff(seq(0, as.numeric(paste0(transcript_length)), 1), df$p5)
    df <- add_missing_rows(df, MISSING, gene_name, transcript_name)
    return(df)
  })
  
  total_big_merged_output <- data.frame()
  sample_names <- names(lst1)
  for (k in seq_along(lst1)) {
    i <- lst1[[k]]
    j <- sample_names[[k]]
    
    # Add a new column to the data frame based on sample name
    i$Sample <- rep(paste0(j), nrow(i))
    
    # Merge the data frame with total_big_merged_output
    total_big_merged_output <- plyr::join(total_big_merged_output, i, type = "full")
  }
  
  total_big_merged_output$Condition <- sapply(str_split(total_big_merged_output$Sample, "_RIBO"), `[`, 1)
  #Now create a new df but with avergae values for each condition.
  p5_pos <- total_big_merged_output$p5
  joined_df <- data.frame(p5=p5_pos)
  for ( sample in unique(total_big_merged_output$Sample)){
    subset_sample <- total_big_merged_output %>% filter(Sample==sample) %>% select(p5, count)
    subset_sample <- subset_sample %>% dplyr::rename(!!paste0(sample):=count)
    joined_df <- join(joined_df,subset_sample, type="inner")
  }
  
  joined_df <- cbind( joined_df, wt_no_se = rowMeans( joined_df[ , grepl("wt", names(joined_df))] ),
                      ko1_no_se = rowMeans( joined_df[ , grepl("ko1", names(joined_df))]))
  
  mean_values <- colnames(joined_df)[c(ncol(joined_df), ncol(joined_df)-1)]
  new_df <- data.frame()
  for (i in mean_values){
    df <- joined_df %>% dplyr::select(p5,i)
    df$Condition <- rep(paste0(i), nrow(df))
    df <- df %>% dplyr::rename("counts"=2)
    df <- df[order(df$p5),]
    df$total <- sum(df$counts)
    df$fraction <- df$counts / df$total
    df$cummulative_counts <- cumsum(df$counts)
    df$cummulative_frac <- cumsum(df$fraction)
    new_df <- rbind(new_df,df)
  }
  
  #Now split df based on selenocystine location.
  five_prime_end <- new_df %>% dplyr::filter(p5 < sec_loc)
  three_prime_end <- new_df %>% dplyr::filter(p5 > sec_loc)
  # custom bins
  data <- 1:sec_loc
  # Number of intervals
  num_intervals <- 5
  # Calculate the width of each interval
  interval_width <- ceiling(length(data) / num_intervals)
  # Calculate the round breakpoints
  break_points <- seq(from = min(data), by = interval_width, length.out = num_intervals + 1)
  # Apply the cut function with the round breakpoints
  bins_five_prime <- cut(data, breaks = break_points, dig.lab = 5)
  level_bins_five_prime <- levels(bins_five_prime)
  # printing the subsets of dataframe
  five_prime_percentages <- data.frame()
  for(i in 1:length(level_bins_five_prime)) {  
    clean_string <- gsub("[^0-9]", " ", levels(bins_five_prime)[i])
    start <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,2]))
    print(start)
    end <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,3]))
    print(end)
    subset_dataframe <- five_prime_end %>% dplyr::filter(between(p5, start, end))
    #now that we have subseted by the bins of interest we will find the top fraction of both conditions and KO-WT
    ko_cond <- subset_dataframe %>% filter(Condition==grep("ko",mean_values, value = T))
    ko_max <- max(ko_cond$cummulative_frac)
    wt_cond <- subset_dataframe %>% filter(Condition==grep("wt",mean_values, value = T))
    wt_max <- max(wt_cond$cummulative_frac)
    value_lfc <- as.data.frame(ko_max-wt_max)
    value_lfc$percentage <- paste0(i)
    value_lfc$region <- paste0(start,":",end)
    value_lfc$footprint <- "five_prime"
    five_prime_percentages <- rbind(five_prime_percentages,value_lfc)
  }
  
  five_prime_percentages$percentage_real <- c("-100", "-80", "-60", "-40", "-20")
  
  #Now for three primer. Calculate transcript length-sec position to know the length.
  #Now split df based on selenocystine location.
  # custom bins
  data_threeprime <- 1:(transcript_length-sec_loc)
  data_threeprime <- data_threeprime + sec_loc
  # Calculate the width of each interval
  interval_width_three_prime <- ceiling(length(data_threeprime) / num_intervals)
  # Calculate the round breakpoints
  break_points_three_prime <- seq(from = min(data_threeprime), by = interval_width_three_prime, length.out = num_intervals + 1)
  # Apply the cut function with the round breakpoints
  bins_three_prime <- cut(data_threeprime, breaks = break_points_three_prime, dig.lab = 5)
  level_bins_three_prime <- levels(bins_three_prime)
  # printing the subsets of dataframe
  three_prime_percentages <- data.frame()
  for (i in 1:length(level_bins_three_prime)) {  
    clean_string <- gsub("[^0-9]", " ", levels(bins_three_prime)[i])
    start <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,2]))
    print(start)
    end <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,3]))
    print(end)
    subset_dataframe <- three_prime_end %>% dplyr::filter(between(p5, start, end))
    #now that we have subseted by the bins of interest we will find the top fraction of both conditions and KO-WT
    ko_cond <- subset_dataframe %>% filter(Condition==grep("ko",mean_values, value = T))
    ko_max <- max(ko_cond$cummulative_frac)
    wt_cond <- subset_dataframe %>% filter(Condition==grep("wt",mean_values, value = T))
    wt_max <- max(wt_cond$cummulative_frac)
    value_lfc_three <- as.data.frame(ko_max-wt_max)
    value_lfc_three$percentage <- paste0(i)
    value_lfc_three$region <- paste0(start,":",end)
    value_lfc_three$footprint <- "three_prime"
    three_prime_percentages <- rbind(three_prime_percentages,value_lfc_three)
  }
  
  three_prime_percentages$percentage_real <- c("+20", "+40", "+60", "+80", "+100")
  
  
  all_together_one_seleno <- rbind(five_prime_percentages, three_prime_percentages)
  all_together_one_seleno$gene <- rep(paste0(gene_name), nrow(all_together_one_seleno))
  
  return(all_together_one_seleno)
  rm(list = ls())



}
```

#Use function in all selenoproteins and bind dfs.
```{r, echo=F, message=F, warning=F}
### GPX1
heatmap_df1 <- heatmap_values_seleno("NM_000581.4","GPX1",899,220)
### GPX3

heatmap_df2 <- heatmap_values_seleno("NM_002084.5","GPX3",1603,281)

### GPX4

heatmap_df3 <- heatmap_values_seleno("NM_002085.5","GPX4",851,267)

### TXNRD1
heatmap_df4 <- heatmap_values_seleno("NM_001093771.3","TXNRD1",3860,1966)

### TXNRD2
heatmap_df5 <- heatmap_values_seleno("NM_006440.5","TXNRD2",1941,1582)

### MSRB1
heatmap_df6 <- heatmap_values_seleno("NM_016332.4","MSRB1",1277,326)

### SELENOH
heatmap_df7 <- heatmap_values_seleno("NM_170746.4","SELENOH",1192,235)
### SELENOI
heatmap_df8 <- heatmap_values_seleno("NM_033505.4","SELENOI",8066,1249)

### SELENOK
heatmap_df9 <- heatmap_values_seleno("NM_021237.5","SELENOK",1497,345)

### SELENON
heatmap_df10 <- heatmap_values_seleno("NM_020451.3","SELENON",4314,417)
### SELENOO
heatmap_df11 <- heatmap_values_seleno("NM_031454.2","SELENOO",2283,2025)

### SELENOS
heatmap_df12 <- heatmap_values_seleno("NM_018445.6","SELENOS",1218,360)

### SELENOT
heatmap_df13 <- heatmap_values_seleno("NM_016275.5","SELENOT",3437,187)

### SELENOW
heatmap_df14 <- heatmap_values_seleno("NM_003009.4","SELENOW",758,120)

### SELENOF
heatmap_df15 <- heatmap_values_seleno("NM_004261.5","SELENOF",1542,301)

### SEPHS2
heatmap_df16 <- heatmap_values_seleno("NM_012248.4","SEPHS2",2244,325)

### TXNRD3
heatmap_df17 <- heatmap_values_seleno("NM_052883.3","TXNRD3",2921,2058)
```

```{r}
rm(all_heatmap_df)
df_list <-  mget(ls(pattern="heatmap_df"))
all_heatmap_df <- do.call(rbind, df_list)
write.csv(all_heatmap_df, "/SSD/maa7095/RIBOseq/downstream_analysis/Coverage_plots/values_for_cov_by_bin_prop/No_sel_cummulative_fraction_by_bin_prop.csv")
```

```{r}
subset_df <- all_heatmap_df %>% dplyr::select(`ko_max - wt_max`, percentage_real, gene)
names(subset_df) <- c("frac_change", "percentage_real", "gene")
top_genes <- subset_df %>% filter(percentage_real=="-20") %>% arrange(desc(frac_change))
order_genes <- top_genes$gene 

col_order <- c("-100", "-80", "-60", "-40", "-20", "+20",  "+40",  "+60",  "+80","+100")

subset_df_mat <- acast(subset_df, gene~percentage_real, value.var = "frac_change")
subset_df_mat <- subset_df_mat[order_genes,col_order]
```

# Heatmap options {.tabset}

## First annotation on the top

```{r, fig.height=8, fig.width=10}
heatmap1 <-Heatmap(subset_df_mat, name = "fold change", cluster_columns =  F,cluster_rows = F ,column_split =rep(c("down","up"), each=5),  column_title_gp = gpar(fill = c("lightcyan", "pink"), font = 1:1), column_gap = unit( 4, "mm"), column_names_rot = 0, column_labels = c("%100", "%80", "%60", "%40", "%20", "%20",  "%40",  "%60",  "%80","%100"))
heatmap1
```

## Second annotation in the bottom

```{r}
column_anno <- HeatmapAnnotation(Sec_pos = c("Down", "Down","Down","Down","Down","Up","Up","Up","Up","Up"),show_annotation_name = FALSE, col = list(Sec_pos = c("Down" = "lightcyan", "Up" = "pink")))
```

```{r, fig.height=8, fig.width=10}
heatmap2 <- Heatmap(subset_df_mat, name = "fold change", cluster_columns =  F,cluster_rows = F ,column_split =rep(1:2, each=5),column_gap = unit( 4, "mm"), column_title = NULL,bottom_annotation = column_anno, column_labels = c("%100", "%80", "%60", "%40", "%20", "%20",  "%40",  "%60",  "%80","%100")) 
heatmap2
```


## {-}

```{r}
sessionInfo()
```

