---
title: "p,a site codon usage publication"
output: html_document
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
---

```{r, echo=F, message=F}
library(devtools)
library(riboWaltz)
library(GenomicFeatures)
library(stringr)
library(plyr)
library(ggplot2)
```

```{r , echo=F, message=F, warning=F}
#Load mapped bam files for each sample, as well as annotation file and trancritpome reference in fasta format.
annotation_db <- create_annotation(gtfpath = "/local/storage/maa7095/references/human/refseq/refseq_v40/GCF_000001405.40_GRCh38.p14_genomic.gtf")
```


```{r , echo=F, message=F, warning=F}
reads_list <- bamtolist(bamfolder = "/SSD/maa7095/RIBOseq/Process/Raw_data/RAW_FASTQ_RIBO_DATA/trimmed_out/non_coding_mapping/kallisto_output_kmer21/maped_bam_files/sorted_bam", annotation = annotation_db)
```


```{r, echo=F, message=F, warning=F}
#Filter reads by length.
filtered_list <- length_filter(data = reads_list,
 			        length_filter_mode = "custom",
 				length_range = 25:50)
```


```{r, echo=F, message=F, warning=F}
#Calculate p-site offset.
psite_offset <- psite(filtered_list, flanking = 6, extremity = "auto", plot = F, plot_dir = "ribowaltz_output_plots")
```


```{r, echo=F, message=F, warning=F}
reads_psite_list <- psite_info(filtered_list, psite_offset, fasta_genome = F,refseq_sep = " ", fastapath = "/local/storage/maa7095/references/human/MANE/MANE.GRCh38.v1.1/MANE.GRCh38.v1.1.refseq_rna.fna",site = c("psite","asite","esite") )
```

#subset selenoproteins
```{r}
selenoprot <- c("NM_000581.4","NM_002083.4","NM_002084.5","NM_002085.5","NM_182701.1","NM_001093771.3","NM_006440.5","NM_013989.5","NM_001362.4","NM_016332.4","NM_170746.4","NM_033505.4","NM_021237.5","NM_080430.4", "NM_020451.3","NM_031454.2","NM_005410.4","NM_018445.6","NM_016275.5","NM_182704.2","NM_003009.4","NM_004261.5","NM_012248.4","NM_052883.3", "NM_000792.7")
```


```{r}
lapply(reads_psite_list, function(df) df[df$transcript %in% selenoprot, ])
```

# P-site analysis.

```{r, echo=F, message=F, warning=F}
codon_list <- unique(reads_psite_list[[1]]$p_site_codon)
scatter_data_frame_p_transcripts <-data.frame(codon=codon_list)
output_folder <-"/SSD/maa7095/RIBOseq/downstream_analysis/Codon_usage/Codon_usage_plots/ALL_transcripts/p_site/" 
for( i in names(reads_psite_list)){

pdf(file = paste0(output_folder,
                  i, "_psite_codon_usage.pdf"),   # The directory you want to save the file in
    height = 12, 
    width = 10) 
 
 codon_usage_barplot <- codon_usage_psite(
   reads_psite_list,
   annotation_db,
   i,
   site = "psite",
   fastapath = "/local/storage/maa7095/references/human/MANE/MANE.GRCh38.v1.1/MANE.GRCh38.v1.1.refseq_rna.fna",
   fasta_genome = F,
   refseq_sep = " ",
   bsgenome = NULL,
   gtfpath = NULL,
   txdb = NULL,
   dataSource = NA,
   organism = NA,
   transcripts = NULL,
   frequency_normalization = TRUE,
   codon_values = NULL,
   label_scatter = FALSE,
   label_number = 64,
   label_aminoacid = T
 )
 
 
 print(codon_usage_barplot[["plot"]])
   
 dev.off()
 values <- as.data.frame(codon_usage_barplot[["dt"]])
 name<-paste0(i)
 name <- str_split(name, "_sorted")[[1]][1]
 name <- str_split(name,"kallisto_non_coding_trimmed_")[[1]][2]
 values <- values %>% dplyr::select(codon,plot_value)
 values <- values %>% dplyr::rename(!!paste0(name):=plot_value)
 scatter_data_frame_p_transcripts <- join(scatter_data_frame_p_transcripts,values,by="codon",type="full")
 
 
}

df_filtered_ALL <- scatter_data_frame_p_transcripts[rowSums(!is.na(scatter_data_frame_p_transcripts[, -1])) > 0, ]
```

```{r, echo=F, message=F, warning=F}
write.table(df_filtered_ALL, "ALL_transcript_psite_codon_usage.csv", quote = F, sep = ",")
```




```{r, echo=F, message=F, warning=F}
#Subset by conditions of interest.
WT_TRANSCRIPTOMES_ALL_df <-df_filtered_ALL %>% dplyr::select(codon,contains("wt"))
WT_TRANSCRIPTOMES_ALL_df$WT_NO_se_MEAN <- rowMeans(WT_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("no_se")))
WT_TRANSCRIPTOMES_ALL_df$WT_SE_MEAN <- rowMeans(WT_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("wt_se")))
#Subset by conditions of interest.
KO1_TRANSCRIPTOMES_ALL_df <-df_filtered_ALL %>% dplyr::select(codon,contains("KO1"))
KO1_TRANSCRIPTOMES_ALL_df$KO1_NO_se_MEAN <- rowMeans(KO1_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("no_se")))
KO1_TRANSCRIPTOMES_ALL_df$KO1_SE_MEAN <- rowMeans(KO1_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("ko1_se")))
#Combine both dataframe, to play with all the values. 
big_df <- join(WT_TRANSCRIPTOMES_ALL_df,KO1_TRANSCRIPTOMES_ALL_df, type="inner")
```



```{r, echo=F, message=F, warning=F}
start_codon <- "AUG"
stop_codon <- c("UAA", "UAG")
sec <- "UGA"
```


```{r, echo=F, message=F, warning=F}
print_plot <- function(ouput_folder, dataframe,x_value,y_value, condition_x, condition_y){
  bs <- 30
  pdf(file = paste0(ouput_folder,condition_x,"_", condition_y,"_SCATTER_psite_codon_usage.pdf"),   
      width = 10,
      height = 12) # The height of the plot in inches
correlation <- round(cor(x_value, y_value), 3)
slope <- round(coef(lm(x_value~y_value))[[2]], 3)
intercept <- round(coef(lm(x_value~y_value))[[1]], 3)
codon_list <- unique(dataframe$codon)
codon_list <- subset(codon_list, !codon_list %in% c("AUG", "UAA","UAG","UGA"))

p<- ggplot(dataframe, aes(x=x_value, y=y_value, colour=codon)) + 
  geom_point()+
  geom_text(label=dataframe$codon, 
    nudge_x = 0.01, nudge_y = 0.02, 
    check_overlap = T) +
  geom_smooth(method = "lm", se=T, color="gray80", fill="gray80", linetype = 1, formula = y ~ x, level = 0.99,  fullrange = TRUE) +
  geom_point(alpha = 0.9, size = 30 * 0.14) +
      scale_colour_manual(name="",values=c("AUG"="blue","UAA"="red", "UAG"="red", "UGA"="orange", codon_list="black")) +
      theme_bw(base_size = 30) +
      theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
      theme(panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) +
      scale_x_continuous(name = paste0("P site all transcripts ", condition_x), limits = c(-0.3,1.3), breaks = c(0,0.25,0.5,0.75,1), expand = c(0,0)) +
      scale_y_continuous(name = paste0("P site all transcriptss ", condition_y), limits = c(-0.3,1.3), breaks = c(0,0.25,0.5,0.75,1), expand = c(0,0)) +
      coord_cartesian(xlim = c(-0.05,1.05), ylim = c(-0.05,1.05)) +
      annotate("text", x = 1, y = c(0, 0.05,0.1), label = c(paste0("R=",correlation), paste0("m=",slope), paste0("b=",intercept)), vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black")
print(p)
dev.off()
return(p)
}
```

## Codon usage under the different conditions. {.tabset}


### WT no SE vs WT SE codon usage.

```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder,big_df, big_df$WT_NO_se_MEAN,big_df$WT_SE_MEAN,"WT no Se Suppl", "WT Se Suppl" )
```

### KO1 no SE vs KO1 SE codon usage.


```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder,big_df, big_df$KO1_NO_se_MEAN,big_df$KO1_SE_MEAN,"KO1 no Se Suppl", "KO1 Se Suppl" )
```

### WT no SE vs KO1 no SE codon usage.
```{r,message=F,echo=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder,big_df, big_df$WT_NO_se_MEAN,big_df$KO1_NO_se_MEAN,"WT no Se Suppl", "KO1 no Se Suppl" )
```

### WT SE vs KO1 SE codon usage.
```{r,message=F, echo=F,fig.width=14, fig.height=12, warning=F}
print_plot(output_folder,big_df, big_df$WT_SE_MEAN,big_df$KO1_SE_MEAN,"WT Se Suppl", "KO1 Se Suppl" )
```


## {-}



```{r, echo=F, message=F, warning=F}
#run p site codon usage based on SElenoproteins only
codon_list <- unique(reads_psite_list[[1]]$p_site_codon)
scatter_data_frame_p_seleno <-data.frame(codon=codon_list)
output_folder <-"/SSD/maa7095/RIBOseq/downstream_analysis/Codon_usage/Codon_usage_plots/ALL_selenoproteins/p_site/" 
for( i in names(reads_psite_list)){

pdf(file = paste0(output_folder,
                  i, "_psite_codon_usage.pdf"),   # The directory you want to save the file in       width = 1000,
       height = 12, 
     width = 10) # The height of the plot in inches
 
 codon_usage_barplot <- codon_usage_psite(
   reads_psite_list,
   annotation_db,
   i,
   site = "psite",
   fastapath = "/local/storage/maa7095/references/human/MANE/MANE.GRCh38.v1.1/MANE.GRCh38.v1.1.refseq_rna.fna",
   fasta_genome = F,
   refseq_sep = " ",
   bsgenome = NULL,
   gtfpath = NULL,
   txdb = NULL,
   dataSource = NA,
   organism = NA,
   transcripts = c("NM_000581.4","NM_002083.4","NM_002084.5","NM_002085.5","NM_182701.1","NM_001093771.3","NM_006440.5","NM_013989.5","NM_001362.4","NM_016332.4","NM_170746.4","NM_033505.4","NM_021237.5","NM_080430.4", "NM_020451.3","NM_031454.2","NM_005410.4","NM_018445.6","NM_016275.5","NM_182704.2","NM_003009.4","NM_004261.5","NM_012248.4","NM_052883.3", "NM_000792.7"),
   frequency_normalization = TRUE,
   codon_values = NULL,
   label_scatter = FALSE,
   label_number = 64,
   label_aminoacid = T
 )
 
 
 print(codon_usage_barplot[["plot"]])
   
 dev.off()
 values <- as.data.frame(codon_usage_barplot[["dt"]])
 write.table(values, paste0(output_folder,
                  i, "_psite_codon_usage.csv"), quote = F, sep = "," )
 name<-paste0(i)
 name <- str_split(name, "_sorted")[[1]][1]
 name <- str_split(name,"kallisto_non_coding_trimmed_")[[1]][2]
 values <- values %>% dplyr::select(codon,plot_value)
 values <- values %>% dplyr::rename(!!paste0(name):=plot_value)
 scatter_data_frame_p_seleno <- join(scatter_data_frame_p_seleno,values,by="codon",type="full")
 
 
}

df_filtered <- scatter_data_frame_p_seleno[rowSums(!is.na(scatter_data_frame_p_seleno[, -1])) > 0, ]
```

```{r, echo=F, message=F, warning=F}
write.table(df_filtered, "Seleno_psite_codon_usage.csv", quote = F, sep = ",")
```




```{r, echo=F, message=F, warning=F}
#Subset by conditions of interest.
WT_SELENOPROTEINS_ALL_df <-df_filtered %>% dplyr::select(codon,contains("wt"))
WT_SELENOPROTEINS_ALL_df$WT_NO_se_MEAN <- rowMeans(WT_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("no_se")))
WT_SELENOPROTEINS_ALL_df$WT_SE_MEAN <- rowMeans(WT_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("wt_se")))
```


```{r, echo=F, message=F, warning=F}
#Subset by conditions of interest.
KO1_SELENOPROTEINS_ALL_df <-df_filtered %>% dplyr::select(codon,contains("KO1"))
KO1_SELENOPROTEINS_ALL_df$KO1_NO_se_MEAN <- rowMeans(KO1_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("no_se")))
KO1_SELENOPROTEINS_ALL_df$KO1_SE_MEAN <- rowMeans(KO1_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("ko1_se")))
```


```{r, echo=F, message=F, warning=F}
#Combine both dataframe, to play with all the values. 
big_seleno_df <- join(WT_SELENOPROTEINS_ALL_df,KO1_SELENOPROTEINS_ALL_df, type="inner")
```


## Codon usage with Selenoprotein subset. {.tabset}

### WT no SE vs WT SE codon usage.

```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$WT_NO_se_MEAN,big_seleno_df$WT_SE_MEAN, "WT no Se suppl","WT Se Suppl" )
```

### KO1 no SE vs KO1 SE codon usage.


```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$KO1_NO_se_MEAN,big_seleno_df$KO1_SE_MEAN, "KO1 no Se Suppl","KO1 Se Suppl" )
```

### WT no SE vs KO1 no SE codon usage.


```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$WT_NO_se_MEAN,big_seleno_df$KO1_NO_se_MEAN, "WT no Se Suppl","KO1 no Se Suppl" )
```

### WT SE vs KO1 SE codon usage.

```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$WT_SE_MEAN,big_seleno_df$KO1_SE_MEAN, "WT Se Suppl","KO1 Se Suppl" )
```

## {-}



# A-site analysis.

```{r, echo=F, message=F, warning=F}
codon_list <- unique(reads_psite_list[[1]]$a_site_codon)
scatter_data_frame_a_transcripts <-data.frame(codon=codon_list)
output_folder <-"/SSD/maa7095/RIBOseq/downstream_analysis/Codon_usage/Codon_usage_plots/ALL_transcripts/a_site/" 
for( i in names(reads_psite_list)){
  
  pdf(file = paste0(output_folder,
                    i, "_asite_codon_usage.pdf"),   # The directory you want to save the file in
      height = 12, 
      width = 10) 
  
  codon_usage_barplot <- codon_usage_psite(
    reads_psite_list,
    annotation_db,
    i,
    site = "asite",
    fastapath = "/local/storage/maa7095/references/human/MANE/MANE.GRCh38.v1.1/MANE.GRCh38.v1.1.refseq_rna.fna",
    fasta_genome = F,
    refseq_sep = " ",
    bsgenome = NULL,
    gtfpath = NULL,
    txdb = NULL,
    dataSource = NA,
    organism = NA,
    transcripts = NULL,
    frequency_normalization = TRUE,
    codon_values = NULL,
    label_scatter = FALSE,
    label_number = 64,
    label_aminoacid = T
  )
  
  
  print(codon_usage_barplot[["plot"]])
  
  dev.off()
  values <- as.data.frame(codon_usage_barplot[["dt"]])
  name<-paste0(i)
  name <- str_split(name, "_sorted")[[1]][1]
  name <- str_split(name,"kallisto_non_coding_trimmed_")[[1]][2]
  values <- values %>% dplyr::select(codon,plot_value)
  values <- values %>% dplyr::rename(!!paste0(name):=plot_value)
  scatter_data_frame_a_transcripts <- join(scatter_data_frame_a_transcripts,values,by="codon",type="full")
  
  
}

df_filtered_ALL <- scatter_data_frame_a_transcripts[rowSums(!is.na(scatter_data_frame_a_transcripts[, -1])) > 0, ]
```

```{r, echo=F, message=F, warning=F}
write.table(df_filtered_ALL, "ALL_transcript_asite_codon_usage.csv", quote = F, sep = ",")
```


```{r, echo=F, message=F, warning=F}
#Subset by conditions of interest.
WT_TRANSCRIPTOMES_ALL_df <-df_filtered_ALL %>% dplyr::select(codon,contains("wt"))
WT_TRANSCRIPTOMES_ALL_df$WT_NO_se_MEAN <- rowMeans(WT_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("no_se")))
WT_TRANSCRIPTOMES_ALL_df$WT_SE_MEAN <- rowMeans(WT_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("wt_se")))
#Subset by conditions of interest.
KO1_TRANSCRIPTOMES_ALL_df <-df_filtered_ALL %>% dplyr::select(codon,contains("KO1"))
KO1_TRANSCRIPTOMES_ALL_df$KO1_NO_se_MEAN <- rowMeans(KO1_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("no_se")))
KO1_TRANSCRIPTOMES_ALL_df$KO1_SE_MEAN <- rowMeans(KO1_TRANSCRIPTOMES_ALL_df %>% dplyr::select(contains("ko1_se")))
#Combine both dataframe, to play with all the values. 
big_df <- join(WT_TRANSCRIPTOMES_ALL_df,KO1_TRANSCRIPTOMES_ALL_df, type="inner")
```


## Codon usage under the different conditions. {.tabset}


### WT no SE vs WT SE codon usage.

```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_df, big_df$WT_NO_se_MEAN,big_df$WT_SE_MEAN,"WT no Se Suppl", "WT Se Suppl" )
```

### KO1 no SE vs KO1 SE codon usage.


```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_df, big_df$KO1_NO_se_MEAN,big_df$KO1_SE_MEAN,"KO1 no Se Suppl", "KO1 Se Suppl" )
```

### WT no SE vs KO1 no SE codon usage.
```{r,message=F,echo=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_df, big_df$WT_NO_se_MEAN,big_df$KO1_NO_se_MEAN,"WT no Se Suppl", "KO1 no Se Suppl" )
```

### WT SE vs KO1 SE codon usage.
```{r,message=F, echo=F,fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_df, big_df$WT_SE_MEAN,big_df$KO1_SE_MEAN,"WT Se Suppl", "KO1 Se Suppl" )
```


## {-}



```{r, echo=F, message=F, warning=F}
#Run a site usage but only for selenoprotein transcripts

codon_list <- unique(reads_psite_list[[1]]$a_site_codon)
scatter_data_frame_a_seleno <-data.frame(codon=codon_list)
output_folder <-"/SSD/maa7095/RIBOseq/downstream_analysis/Codon_usage/Codon_usage_plots/ALL_selenoproteins/a_site/" 
for( i in names(reads_psite_list)){
  
  pdf(file = paste0(output_folder,
                    i, "_asite_codon_usage.pdf"),   # The directory you want to save the file in       width = 1000,
      height = 12, 
      width = 10) # The height of the plot in inches
  
  codon_usage_barplot <- codon_usage_psite(
    reads_psite_list,
    annotation_db,
    i,
    site = "asite",
    fastapath = "/local/storage/maa7095/references/human/MANE/MANE.GRCh38.v1.1/MANE.GRCh38.v1.1.refseq_rna.fna",
    fasta_genome = F,
    refseq_sep = " ",
    bsgenome = NULL,
    gtfpath = NULL,
    txdb = NULL,
    dataSource = NA,
    organism = NA,
    transcripts = c("NM_000581.4","NM_002083.4","NM_002084.5","NM_002085.5","NM_182701.1","NM_001093771.3","NM_006440.5","NM_013989.5","NM_001362.4","NM_016332.4","NM_170746.4","NM_033505.4","NM_021237.5","NM_080430.4", "NM_020451.3","NM_031454.2","NM_005410.4","NM_018445.6","NM_016275.5","NM_182704.2","NM_003009.4","NM_004261.5","NM_012248.4","NM_052883.3", "NM_000792.7"),
    frequency_normalization = TRUE,
    codon_values = NULL,
    label_scatter = FALSE,
    label_number = 64,
    label_aminoacid = T
  )
  
  
  print(codon_usage_barplot[["plot"]])
  
  dev.off()
  values <- as.data.frame(codon_usage_barplot[["dt"]])
  write.table(values, paste0(output_folder,
                  i, "_asite_codon_usage.csv"), quote = F, sep = "," )
  name<-paste0(i)
  name <- str_split(name, "_sorted")[[1]][1]
  name <- str_split(name,"kallisto_non_coding_trimmed_")[[1]][2]
  values <- values %>% dplyr::select(codon,plot_value)
  values <- values %>% dplyr::rename(!!paste0(name):=plot_value)
  scatter_data_frame_a_seleno <- join(scatter_data_frame_a_seleno,values,by="codon",type="full")
  
  
}

df_filtered <- scatter_data_frame_a_seleno[rowSums(!is.na(scatter_data_frame_a_seleno[, -1])) > 0, ]
```

```{r, echo=F, message=F, warning=F}
write.table(df_filtered, "Seleno_asite_codon_usage.csv", quote = F, sep = ",")
```

```{r, echo=F, message=F, warning=F}
#Subset by conditions of interest.
WT_SELENOPROTEINS_ALL_df <-df_filtered %>% dplyr::select(codon,contains("wt"))
WT_SELENOPROTEINS_ALL_df$WT_NO_se_MEAN <- rowMeans(WT_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("no_se")))
WT_SELENOPROTEINS_ALL_df$WT_SE_MEAN <- rowMeans(WT_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("wt_se")))
#Subset by conditions of interest.
KO1_SELENOPROTEINS_ALL_df <-df_filtered %>% dplyr::select(codon,contains("KO1"))
KO1_SELENOPROTEINS_ALL_df$KO1_NO_se_MEAN <- rowMeans(KO1_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("no_se")))
KO1_SELENOPROTEINS_ALL_df$KO1_SE_MEAN <- rowMeans(KO1_SELENOPROTEINS_ALL_df %>% dplyr::select(contains("ko1_se")))
#Combine both dataframe, to play with all the values. 
big_seleno_df <- join(WT_SELENOPROTEINS_ALL_df,KO1_SELENOPROTEINS_ALL_df, type="inner")
```



## Codon usage with Selenoprotein subset. {.tabset}

### WT no SE vs WT SE codon usage.

```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$WT_NO_se_MEAN,big_seleno_df$WT_SE_MEAN, "WT no Se suppl","WT Se Suppl" )
```

### KO1 no SE vs KO1 SE codon usage.


```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$KO1_NO_se_MEAN,big_seleno_df$KO1_SE_MEAN, "KO1 no Se Suppl","KO1 Se Suppl" )
```

### WT no SE vs KO1 no SE codon usage.


```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$WT_NO_se_MEAN,big_seleno_df$KO1_NO_se_MEAN, "WT no Se Suppl","KO1 no Se Suppl" )
```

### WT SE vs KO1 SE codon usage.

```{r, echo=F, message=F, fig.width=14, fig.height=12, warning=F}
print_plot(output_folder, big_seleno_df, big_seleno_df$WT_SE_MEAN,big_seleno_df$KO1_SE_MEAN, "WT Se Suppl","KO1 Se Suppl" )
```

## {-}


# Get codon usage for each selenoprotein separately.
```{r, echo=F, message=F}
codon_list <- (big_seleno_df$codon)
seleno_individual_usage <- function(seleno_id,seleno_name,x_site, output_folder){
seleno_ind_codon_p_site <-data.frame(codon=codon_list)
for( i in names(reads_psite_list)){
codon_usage_barplot <- codon_usage_psite(
  reads_psite_list,
  annotation_db,
  i,
  site = x_site,
  fastapath = "/local/storage/maa7095/references/human/MANE/MANE.GRCh38.v1.1/MANE.GRCh38.v1.1.refseq_rna.fna",
  fasta_genome = F,
  refseq_sep = " ",
  bsgenome = NULL,
  gtfpath = NULL,
  txdb = NULL,
  dataSource = NA,
  organism = NA,
  transcripts = seleno_id,
  frequency_normalization = TRUE,
  codon_values = NULL,
  label_scatter = FALSE,
  label_number = 64,
  label_aminoacid = T
)


values <- as.data.frame(codon_usage_barplot[["dt"]])

name<-paste0(i)
name <- str_split(name, "_sorted")[[1]][1]
name <- str_split(name,"kallisto_non_coding_trimmed_")[[1]][2]
values <- values %>% dplyr::select(codon,raw_value,normalized_value)
values <- values %>% dplyr::rename(!!paste0(name, "_raw"):=raw_value, !!paste0(name, "_normalized"):=normalized_value)
seleno_ind_codon_p_site <- join(seleno_ind_codon_p_site,values,by="codon",type="full")

}

seleno_ind_codon_p_site[is.na(seleno_ind_codon_p_site)] <- 0
write.table(seleno_ind_codon_p_site, paste0(output_folder,seleno_name,"_",
                                name,"_",x_site, "_codon_usage.csv"), quote = F, sep = ",", row.names = F )
}
```


```{r, echo=F, message=F}
#output directory
pout_folder <-"/SSD/maa7095/RIBOseq/downstream_analysis/Codon_usage/Seleno_separated/p_site/" 

seleno_individual_usage("NM_000581.4","GPX1","psite", pout_folder)

seleno_individual_usage("NM_002084.5","GPX3","psite", pout_folder)

seleno_individual_usage("NM_002085.5","GPX4","psite", pout_folder)

seleno_individual_usage("NM_001093771.3","TXNRD1","psite", pout_folder)

seleno_individual_usage("NM_006440.5","TXNRD2","psite", pout_folder)

seleno_individual_usage("NM_016332.4","MSRB1","psite", pout_folder)

seleno_individual_usage("NM_170746.4","SELENOH","psite", pout_folder)

seleno_individual_usage("NM_033505.4","SELENOI","psite", pout_folder)

seleno_individual_usage("NM_021237.5","SELENOK","psite", pout_folder)

seleno_individual_usage("NM_020451.3","SELENON","psite", pout_folder)

seleno_individual_usage("NM_031454.2","SELENOO","psite", pout_folder)

seleno_individual_usage("NM_018445.6","SELENOS","psite", pout_folder)

seleno_individual_usage("NM_016275.5","SELENOT","psite", pout_folder)

seleno_individual_usage("NM_003009.4","SELENOW","psite", pout_folder)

seleno_individual_usage("NM_004261.5","SELENOF","psite", pout_folder)

seleno_individual_usage("NM_012248.4","SEPHS2","psite", pout_folder)

seleno_individual_usage("NM_052883.3","TXNRD3","psite", pout_folder)

seleno_individual_usage("NM_000792.7","DIO1","psite", pout_folder)

```

```{r, echo=F, message=F}
#output directory
aout_folder <-"/SSD/maa7095/RIBOseq/downstream_analysis/Codon_usage/Seleno_separated/a_site/" 

seleno_individual_usage("NM_000581.4","GPX1","asite", aout_folder)

seleno_individual_usage("NM_002084.5","GPX3","asite", aout_folder)

seleno_individual_usage("NM_002085.5","GPX4","asite", aout_folder)

seleno_individual_usage("NM_001093771.3","TXNRD1","asite", aout_folder)

seleno_individual_usage("NM_006440.5","TXNRD2","asite", aout_folder)

#seleno_individual_usage("NM_012331.5","MSRA","asite", aout_folder)

seleno_individual_usage("NM_016332.4","MSRB1","asite", aout_folder)

seleno_individual_usage("NM_170746.4","SELENOH","asite", aout_folder)

seleno_individual_usage("NM_033505.4","SELENOI","asite", aout_folder)

seleno_individual_usage("NM_021237.5","SELENOK","asite", aout_folder)

seleno_individual_usage("NM_020451.3","SELENON","asite", aout_folder)

seleno_individual_usage("NM_031454.2","SELENOO","asite", aout_folder)

seleno_individual_usage("NM_018445.6","SELENOS","asite", aout_folder)

seleno_individual_usage("NM_016275.5","SELENOT","asite", aout_folder)

seleno_individual_usage("NM_003009.4","SELENOW","asite", aout_folder)

seleno_individual_usage("NM_004261.5","SELENOF","asite", aout_folder)

seleno_individual_usage("NM_012248.4","SEPHS2","asite", aout_folder)

seleno_individual_usage("NM_052883.3","TXNRD3","asite", aout_folder)

seleno_individual_usage("NM_000792.7","DIO1","asite", aout_folder)

```


