---
title: "Footprints analysis publication"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    df_print: paged
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',tidy=TRUE, cache=FALSE, echo = F)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```


```{r, echo=F, message=F}
#Load libraries
library(stringr)
library(plyr)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(grid)
library(tidyverse)
library(rstatix)
```

```{r, echo=F, message=F}
#Read the output of the coverage script'
#transcript name: MANE trancript name
#gene name: gene name
#transcript length: transcript length 
#sec_loc: selenocytine position at the transcript
```

The 5' and 3' prime footprints were determined based on read counts obtained from the CDF analysis. Using a fixed 200 bp window, we identified the regions corresponding to the 5' and 3' ends. Specifically, reads from the Sec codon position and below 200 bp were considered as 5' prime footprints, while reads from the Sec position and above 200 bp were considered as 3' prime footprints.

For each sample, we summed the reads within these respective regions. Then, for each condition, we calculated the average values of the footprints across the samples.

Regarding RNA abundance, we also obtained reads for each position in the transcript. To focus on the relevant information, we subsetted the values to a 400 bp window. Within this window, positions below 200 bp and above 200 bp to Sec position.

The comparisons we performed were:

3' prime / 5' prime: This ratio quantifies the relative abundance of 3' prime footprints compared to 5' prime footprints.

```{r, echo=F, message=F}
footprint_df <- function(transcript_name,gene_name,transcript_length,sec_loc, data_type, window_sec){
  my_files <- list.files(paste0(cover_dir, "/",transcript_name), pattern="_no_se.*\\.cov")
  my_files2 <- str_extract(my_files, '.*(?=\\.cov)')
  #All the files in same format to read it correctly
  for (i in my_files2) {
    filename <- str_split(i, "_trimmed_")[[1]][2]
    wd <- paste0(cover_dir, "/", transcript_name, "/", i, ".cov")
    if (file.size(wd) > 0) {
      assign(filename, read.csv(wd, header=T, sep="\t"))
    } else {
      cat("File", wd, "is empty.\n")
    }
  }
  
  #Change the names of prob by each samples name.
  lst <-  mget(ls(pattern=paste0(data_type)))
  lst1 <- Map(function(x,y) {names(x)[1] <- y; x},  lst, paste0(gene_name))
  lst1 <- Map(function(x,y) {names(x)[2] <- y; x},  lst1, "p5")
  lst1 <- Map(function(x,y) {names(x)[3] <- y; x},  lst1, "count")
  
  
  list2env(lst1, envir=.GlobalEnv)
  
  total_rows <- sum(unlist(lapply(lst1, function(df) dim(df)[1])))
  total_col <- sum(unlist(lapply(lst1, function(df) dim(df)[2])))
  
  add_missing_rows <- function(df, MISSING, transcript_name, gene_name) {
    df <- rbind(df,setNames(data.frame(paste0(gene_name), MISSING, 0), c(paste(transcript_name), "p5", "count")))
    df[[""]] <- transcript_name
    df <- df[order(df$p5),]
    return(df)
  }
  
  lst1 <- lapply(lst1, function(df) {
    MISSING <- setdiff(seq(0, as.numeric(paste0(transcript_length)), 1), df$p5)
    df <- add_missing_rows(df, MISSING, gene_name, transcript_name)
    return(df)
  })
  
  total_big_merged_output <- data.frame()
  sample_names <- names(lst1)
  for (k in seq_along(lst1)) {
    i <- lst1[[k]]
    j <- sample_names[[k]]
    
    # Add a new column to the data frame based on sample name
    i$Sample <- rep(paste0(j), nrow(i))
    
    # Merge the data frame with total_big_merged_output
    total_big_merged_output <- plyr::join(total_big_merged_output, i, type = "full")
  }
  
  total_big_merged_output$Condition <- sapply(str_split(total_big_merged_output$Sample,  paste0("_",data_type)), `[`, 1)
  if(isTRUE(data_type=="RIBO")){
    total_big_merged_output <- total_big_merged_output %>% 
      dplyr::mutate(footprint = ifelse(p5 >= sec_loc, "3_prime", "5_prime"))
  }
  
  #set up a window to calculate coverage through each selenoprotein.
  total_big_merged_output <- total_big_merged_output %>% filter(!(p5 > (sec_loc+window_sec) | p5 < (sec_loc-window_sec)))
  
  #now sum counts from sample sample and 3 and 5 prime.
  footprint_df <- total_big_merged_output %>%
    dplyr::group_by(Sample, footprint) %>%
    dplyr::summarize(sum_counts = sum(count)) %>%
    ungroup()
  
  # Step 1: Calculate total sum_counts for each sample
  total_counts <- footprint_df %>%
    group_by(Sample) %>%
    summarize(total_sum_counts = sum(sum_counts))
  
  footprint_df$Condition <- sapply(str_split(footprint_df$Sample, paste0("_",data_type)), `[`, 1)
  footprint_df$Gene <- rep(gene_name, nrow(footprint_df))
  footprint_df$Data_type <- rep(data_type, nrow(footprint_df))
  return(footprint_df)
}


```

```{r, echo=F, message=F}
#RIBO location
cover_dir <- "/SSD/maa7095/RIBOseq/Process/Raw_data/RAW_FASTQ_RIBO_DATA/trimmed_out/non_coding_mapping/kallisto_output_kmer21/maped_bam_files"
footprint_df_results_list <- list()

footprint_df_results_list[["GPX1"]] <- footprint_df("NM_000581.4","GPX1",899,220, "RIBO", 200)

footprint_df_results_list[["GPX3"]] <- footprint_df("NM_002084.5","GPX3",1603,281, "RIBO", 200)

footprint_df_results_list[["GPX4"]] <- footprint_df("NM_002085.5","GPX4",851,267, "RIBO", 200)

footprint_df_results_list[["TXNRD1"]] <- footprint_df("NM_001093771.3","TXNRD1",3860,1966, "RIBO", 200)

footprint_df_results_list[["TXNRD2"]] <- footprint_df("NM_006440.5","TXNRD2",1941,1582, "RIBO", 200)


footprint_df_results_list[["MSRB1"]] <- footprint_df("NM_016332.4","MSRB1",1277,326, "RIBO", 200)

footprint_df_results_list[["SELENOH"]] <- footprint_df("NM_170746.4","SELENOH",1192,235, "RIBO", 200)

footprint_df_results_list[["SELENOI"]] <- footprint_df("NM_033505.4","SELENOI",8066,1249, "RIBO", 200)

footprint_df_results_list[["SELENOK"]] <- footprint_df("NM_021237.5","SELENOK",1497,345, "RIBO",200)

footprint_df_results_list[["SELENON"]] <- footprint_df("NM_020451.3","SELENON",4314,417, "RIBO", 200)

footprint_df_results_list[["SELENOO"]] <- footprint_df("NM_031454.2","SELENOO",2283,2025, "RIBO", 200)

footprint_df_results_list[["SELENOS"]] <- footprint_df("NM_018445.6","SELENOS",1218,360, "RIBO",200)

footprint_df_results_list[["SELENOT"]] <- footprint_df("NM_016275.5","SELENOT",3437,187, "RIBO", 200)

footprint_df_results_list[["SELENOW"]] <- footprint_df("NM_003009.4","SELENOW",758,120, "RIBO", 100)

footprint_df_results_list[["SELENOF"]] <- footprint_df("NM_004261.5","SELENOF",1542,301, "RIBO", 200)

footprint_df_results_list[["SEPHS2"]]<- footprint_df("NM_012248.4","SEPHS2",2244,325, "RIBO", 200)

footprint_df_results_list[["TXNRD3"]] <- footprint_df("NM_052883.3","TXNRD3",2921,2058, "RIBO", 200)
```


```{r, echo=F, message=F}
complete_fottprint_df <- do.call(rbind,footprint_df_results_list)
complete_fottprint_df$Sample_2 <- complete_fottprint_df$Sample
complete_fottprint_df$Sample_2 <- str_replace_all(complete_fottprint_df$Sample_2, "_RIBO", "")
```


## Frequency bar plots for NO Selenium Suplementation. {.tabset}


```{r, echo=F, message=F}
RIBO_complete_fottprint_df <- complete_fottprint_df %>% filter(Data_type=="RIBO")

#Subset RIBO data.
fraction_RIBO_df <- data.frame()
for (i in unique(RIBO_complete_fottprint_df$Gene)){
  freq_df <- RIBO_complete_fottprint_df %>% filter(Gene==i)
  freq_df_2 <- freq_df %>%
    group_by(Sample) %>%
    filter(footprint %in% c("3_prime", "5_prime")) %>%
    pivot_wider(names_from = footprint, values_from = sum_counts) %>%
    mutate(Fraction = `3_prime` / `5_prime`)
  fraction_RIBO_df <- rbind(fraction_RIBO_df, freq_df_2)
}


fraction_RIBO_df3 <- fraction_RIBO_df %>% arrange(Gene)
stat.test <- fraction_RIBO_df3 %>%
  group_by(Gene) %>%
  t_test(Fraction ~ Condition)
stat.test <- stat.test %>%
  add_xy_position(x = "Gene", dodge = 0.8)

#write.csv(fraction_RIBO_df3, "/SSD/maa7095/RIBOseq/downstream_analysis/Footprints/Footprint_200nt_5_3_RIBO_NoSe.csv")
```





### Bar plot for RIBO fraction 3'/5' footprints for each gene.
```{r, echo=F, message=F}
# Create a bar plot with error bars (mean +/- sd)
pdf(file = "/SSD/maa7095/RIBOseq/downstream_analysis/Footprints/RIBO3_5_NO_Se_supplementation.pdf",   
      width = 12,
      height = 10) # The height of the plot in inches

plot <- ggbarplot(
  fraction_RIBO_df3, x = "Gene", y = "Fraction", add = "mean_sd", 
  color = "Condition", palette = c("salmon1", "slategray"),
  position = position_dodge(0.8)
  ) + stat_pvalue_manual(
  stat.test,  label = "p", tip.length = 0.01, label.size = 3,
  x = "Gene"
  ) + 
  labs(x = "Selenoproteins", 
       y = "3'RPF/5'RPF") + 
  theme(legend.title = element_text(size=12),
        legend.text = element_text(size=10)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(
    # Hide panel borders and remove grid lines
    panel.border = element_blank(),
    panel.background=element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # Change axis line
    axis.line = element_line(colour = "gray"))
print(plot)

dev.off()
plot

```

## {-}
