---
title: "Script 2 cell lines CDF stalling plots"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    df_print: paged
---

```{r, echo=F, message=F, warning=F}
#Load libraries
library(stringr)
library(plyr)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(grid)
```


```{r, echo=F, message=F, warning=F}
#From the output of Script1_ribosome_stalling.Rmd load the counts based on inframe psite coverage for each Selenoprotein.
#transcript name: Selenoprotein MANE v1.2 gencode annotation 
#gene name: gene symbol
#transcript length: transcript length 
#sec_loc: selenocysteine position at the transcript
#start_pos: start codon position in the transcript
#stop_pos: stop codon position in the transcript
#condition: "no_se" or "se"
coverage_plot <- function(transcript_name,gene_name,transcript_length,sec_loc, start_pos, stop_pos, condition){
  #load results from Script 1
cover_dir <- "p_site_coverage/"

  my_files <- list.files(paste0(cover_dir, "/",transcript_name), pattern=".cov")
  my_files2 <- str_extract(my_files, '.*(?=\\.cov)')
  #All the files in same format to read it correctly
  for (i in my_files2) {
    filename <- str_split(i, paste0(transcript_name, "_"))[[1]][2]
    wd <- paste0(cover_dir, "/", transcript_name, "/", i, ".cov")
    if (file.size(wd) > 0) {
      assign(filename, read.csv(wd, header=T, sep="\t"))
    } else {
      cat("File", wd, "is empty.\n")
    }
  }
  
  #Change the names of prob by each samples name.
  lst <-  mget(ls(pattern="RIBO"))
  lst1 <- Map(function(x,y) {names(x)[1] <- y; x},  lst, paste0(gene_name))
  lst1 <- Map(function(x,y) {names(x)[2] <- y; x},  lst1, "p5")
  lst1 <- Map(function(x,y) {names(x)[3] <- y; x},  lst1, "count")
  
  
  list2env(lst1, envir=.GlobalEnv)
  
  total_rows <- sum(unlist(lapply(lst1, function(df) dim(df)[1])))
  #print(total_rows)
  total_col <- sum(unlist(lapply(lst1, function(df) dim(df)[2])))
  #print(total_col)
  
  add_missing_rows <- function(df, MISSING, transcript_name, gene_name) {
    df <- rbind(df,setNames(data.frame(paste0(gene_name), MISSING, 0), c(paste0(transcript_name), "p5", "count")))
    df[[""]] <- transcript_name
    df <- df[order(df$p5),]
    return(df)
  }
  
  lst1 <- lapply(lst1, function(df) {
    MISSING <- setdiff(seq(0, as.numeric(paste0(transcript_length)), 1), df$p5)
    df <- add_missing_rows(df, MISSING, gene_name, transcript_name)
    return(df)
  })
  
  total_big_merged_output <- data.frame()
  sample_names <- names(lst1)
  for (k in seq_along(lst1)) {
    i <- lst1[[k]]
    j <- sample_names[[k]]
    
    # Add a new column to the data frame based on sample name
    i$Sample <- rep(paste0(j), nrow(i))
    
    # Merge the data frame with total_big_merged_output
    total_big_merged_output <- plyr::join(total_big_merged_output, i, type = "full")
  }
  
  total_big_merged_output$Condition <- sapply(str_split(total_big_merged_output$Sample, "_RIBO"), `[`, 1)
  
  total_big_merged_output <- total_big_merged_output %>% dplyr::filter(grepl(condition, Condition))
  if(isTRUE(length(unique(total_big_merged_output$Condition))==4)){
    seleno_samples <- c("ko1_se","wt_se")
    total_big_merged_output <- total_big_merged_output %>% filter(Condition %in%seleno_samples ) 
  }
  p5_pos <- unique(total_big_merged_output$p5)
  joined_df <- data.frame(p5=p5_pos)
  for ( sample in unique(total_big_merged_output$Sample)){
    subset_sample <- total_big_merged_output %>% filter(Sample==sample) %>% dplyr::select(p5, count)
    subset_sample <- subset_sample %>% dplyr::rename(!!paste0(sample):=count)
    joined_df <- join(joined_df,subset_sample, type="inner")
  }
  
  joined_df <- cbind( joined_df, wt = rowMeans( joined_df[ , grepl("wt", names(joined_df))] ),
                      ko1 = rowMeans( joined_df[ , grepl("ko1", names(joined_df))]))
  
  colnames(joined_df)[c(ncol(joined_df), ncol(joined_df)-1)] <- paste0(colnames(joined_df)[c(ncol(joined_df), ncol(joined_df)-1)], "_", condition)
  mean_values <- colnames(joined_df)[c(ncol(joined_df), ncol(joined_df)-1)]
  new_df <- data.frame()
  for (i in mean_values){
    df <- joined_df %>% dplyr::select(p5,i)
    df$Condition <- rep(paste0(i), nrow(df))
    df <- df %>% dplyr::rename("counts"=2)
    df <- df[order(df$p5),]
    df$total <- sum(df$counts)
    df$fraction <- df$counts / df$total
    df$cummulative_counts <- cumsum(df$counts)
    df$cummulative_frac <- cumsum(df$fraction)
    new_df <- rbind(new_df,df)
  }
  write.csv(new_df,paste0("CDF_plots/", condition,   "_by_groups/", gene_name,"_coverage_df.csv"), quote=F, row.names = F)
  
  new_df <- new_df %>%
    mutate(counts = ifelse(str_detect(Condition, "ko1"), -1 * counts, counts))
  
  
  pdf(file = paste0("CDF_plots/", condition, "_by_groups/",gene_name, "_coverage_clean_counts.pdf"),   # The directory you want to save the file in
      width = 10,
      height = 12) # 
  
  max_index <- which.max(new_df$counts)
  max_count <- new_df$counts[max_index]
  ko1_name <- grep("ko1", unique(new_df$Condition), value = T)
  wt_name <- grep("wt", unique(new_df$Condition), value = T)
  color_values <- c(
    setNames("salmon1", ko1_name),
    setNames("slategray", wt_name)
  )
  alpha_values <- c(
    setNames(0.5, ko1_name),
    setNames(0.5, wt_name)
  )
  linetype_values <- c(
    setNames("solid", ko1_name),
    setNames("solid", wt_name)
  )
  
  coverage_plot1 <- ggplot() + 
    geom_segment(data=new_df, 
                 aes(x=p5, 
                     xend=p5, 
                     y=0, yend=counts, 
                     #group = interaction(Condition, Sample), 
                     colour=Condition,
                     alpha=Condition)) +
    #linetype=Condition)) +
    scale_alpha_manual(name = "Condition",
                       values = alpha_values) +
    scale_colour_manual(values =color_values) + 
    geom_point(aes(x=stop_pos, y=0),shape=4, fill="black", color="black", size=4) +
    geom_point(aes(x=start_pos, y=0),shape=4, fill="black", color="black", size=4) +
    geom_segment(aes(x = sec_loc, 
                     y = max_count, 
                     xend = sec_loc, 
                     yend = 0), 
                 linetype="dashed", 
                 colour = "red") +
    geom_text(aes(x=sec_loc, 
                  y= abs(max_count) + 0.5),
              label = "Sec",
              colour = "black", 
              vjust = 1, 
              size = 4) +
    labs(x = paste0(gene_name, " 5' alignment position"), 
         y = "psite coverage") + 
    theme(legend.title = element_text(size=12),
          legend.text = element_text(size=10)) +
    theme(
      # Hide panel borders and remove grid lines
      panel.border = element_blank(),
      panel.background=element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      # Change axis line
      axis.line = element_line(colour = "gray")) +
    geom_hline(yintercept = 0, color = "gray") 
  
  print(coverage_plot1)
  dev.off()
  
  pdf(file = paste0("CDF_plots/", condition, "_by_groups/",gene_name, "_coverage.pdf"),   # The directory you want to save the file in
      width = 10,
      height = 12) 
  coverage_plot <- ggplot() + 
    geom_step(data=new_df, 
              aes(x=p5, 
                  y=cummulative_frac, 
                  group = Condition, 
                  colour=Condition, 
                  size=Condition,
                  linetype = Condition))  +
    scale_linetype_manual(name = "Condition",
                          values = linetype_values) +
    scale_colour_manual(values = color_values) + 
    geom_segment(aes(x = stop_pos, 
                     y = 0, 
                     xend = stop_pos, 
                     yend = 1), 
                 linetype="dashed", 
                 colour = "black") +
    geom_segment(aes(x = start_pos, 
                     y = 0, 
                     xend = start_pos, 
                     yend = 1), 
                 linetype="dashed", 
                 colour = "black") +
    geom_segment(aes(x = sec_loc, 
                     y = 0, 
                     xend = sec_loc, 
                     yend = max(new_df$cummulative_frac)), 
                 linetype="dashed", 
                 colour = "red") +
    geom_text(aes(x=sec_loc, 
                  y= max(new_df$cummulative_frac) + 0.05),
              label = "Sec",
              colour = "black", 
              vjust = 1, 
              size = 4) + 
    labs(x = paste0(gene_name, " 5' alignment position"), 
         y = "Fraction of psite") + 
    scale_size_manual(values = alpha_values) +
    theme(legend.title = element_text(size=12),
          legend.text = element_text(size=10)) +
    theme(
      # Hide panel borders and remove grid lines
      panel.border = element_blank(),
      panel.background=element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      # Change axis line
      axis.line = element_line(colour = "gray")
    )   
  print(coverage_plot)
  
  
  dev.off()
  plot <- ggarrange(coverage_plot1,coverage_plot,
                    ncol = 1, nrow = 2)
  
  return(plot)
  rm(list = ls())
  
  
}
```

## Coverage plots for all Selenoproteins.{.tabset}

### GPX1
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000419783.3","GPX1",899,220, 76, 685, "no_se")
```

### GPX3
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000388825.9","GPX3",1603,281, 65, 743, "no_se")
```


### GPX4
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000354171.13","GPX4",851,267, 51, 642, "no_se")
```

### TXNRD1
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000525566.6","TXNRD1",3860,1966, 25, 1972, "no_se")
```

### TXNRD2
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000400521.7","TXNRD2",1941,1582, 16, 1588, "no_se")
```

### MSRB1
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000361871.8","MSRB1",1277,326, 44, 392, "no_se")
```

### SELENOH
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000534355.6","SELENOH",1192,235, 106, 472, "no_se")
```

### SELENOI
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000260585.12","SELENOI",8066,1249, 91, 1282, "no_se")
```

### SELENOK
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000495461.6","SELENOK",1497,345, 72, 354, "no_se")
```

### SELENON
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000361547.7","SELENON",4314,417, 39, 1707, "no_se")
```


### SELENOO
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000380903.7","SELENOO",2283,2025, 27, 2034, "no_se")
```

### SELENOS
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000526049.6","SELENOS",1218,360, 69, 636, "no_se")
```

### SELENOT
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000471696.6","SELENOT",3437,187, 43, 628, "no_se")
```

### SELENOW
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000601048.6","SELENOW",758,120, 84, 345, "no_se")
```

### SELENOF
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000331835.10","SELENOF",1542,301, 16, 511, "no_se")
```

### SEPHS2
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000478753.5","SEPHS2",2244,325, 148, 1492, "no_se")
```

### TXNRD3
```{r, echo=F, message=F, warning=F}
coverage_plot("ENST00000524230.9","TXNRD3",2921,2058, 135, 2064, "no_se")
```




## {-}
