---
title: "Script 3 PDX Heatmaps for ribosome stalling"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    df_print: paged
---

For this heatmap: from the in-frame p-site coverages obtained from Script1 we are able to calculate the cummulative fraction for each selenoprotein. Using cummulative values we can calculate the differences between Metastasis (Met_liv) - Pirmary tumor (SQ) for fixed bins up and down to Sec to Start codon position and Stop codon position. How? We identified 20 % windows (100, 80, 60, 40, 20) from CDS start to Sec and from Sec to stop codon. And on those windows we use the max cummulative fractions for each Met_liv and sq and calculate difference Met_liv - sq. 

```{r, echo=F, message=F, warning=F}
#Load libraries
library(stringr)
library(plyr)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(grid)
library(decoupleR)
library(reshape2)
library(ComplexHeatmap)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE, echo = F)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```


```{r, echo=F, message=F, warning=F}
#From the output of Script1_ribosome_stalling.Rmd load the counts based on inframe psite coverage for each Selenoprotein.
#transcript name: Selenoprotein MANE v1.2 gencode annotation 
#gene name: gene symbol
#transcript length: transcript length 
#sec_loc: selenocysteine position at the transcript
#condition: "no_se" or "se"
#cds_start: start codon position in the transcript
#cds_stop: stop codon position in the transcript
heatmap_values_seleno <- function(transcript_name,gene_name,transcript_length,sec_loc, cds_start, cds_stop){
  #load results from Script 1
  cover_dir <- "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Xenograft/p_site_coverage"
  
  my_files <- list.files(paste0(cover_dir, "/",transcript_name), pattern=".cov")
  my_files2 <- str_extract(my_files, '.*(?=\\.cov)')
  #All the files in same format to read it correctly
  for (i in my_files2) {
    filename <- str_split(i, paste0(transcript_name, "_"))[[1]][2]
    wd <- paste0(cover_dir, "/", transcript_name, "/", i, ".cov")
    if (file.size(wd) > 0) {
      assign(filename, read.csv(wd, header=T, sep="\t"))
    } else {
      cat("File", wd, "is empty.\n")
    }
  }
  
  #Change the names of prob by each samples name.
  lst <-  mget(ls(pattern="cm2"))
  lst1 <- Map(function(x,y) {names(x)[1] <- y; x},  lst, paste0(gene_name))
  lst1 <- Map(function(x,y) {names(x)[2] <- y; x},  lst1, "p5")
  lst1 <- Map(function(x,y) {names(x)[3] <- y; x},  lst1, "count")
  
  
  list2env(lst1, envir=.GlobalEnv)
  
  total_rows <- sum(unlist(lapply(lst1, function(df) dim(df)[1])))
  total_col <- sum(unlist(lapply(lst1, function(df) dim(df)[2]))) 
  
  add_missing_rows <- function(df, MISSING, transcript_name, gene_name) {
    df <- rbind(df,setNames(data.frame(paste0(gene_name), MISSING, 0), c(paste0(transcript_name), "p5", "count")))
    df[[""]] <- transcript_name
    df <- df[order(df$p5),]
    return(df)
  }
  
  lst1 <- lapply(lst1, function(df) {
    MISSING <- setdiff(seq(0, as.numeric(paste0(transcript_length)), 1), df$p5)
    df <- add_missing_rows(df, MISSING, gene_name, transcript_name)
    return(df)
  })
  
  total_big_merged_output <- data.frame()
  sample_names <- names(lst1)
  for (k in seq_along(lst1)) {
    i <- lst1[[k]]
    j <- sample_names[[k]]
    
    # Add a new column to the data frame based on sample name
    i$Sample <- rep(paste0(j), nrow(i))
    
    # Merge the data frame with total_big_merged_output
    total_big_merged_output <- plyr::join(total_big_merged_output, i, type = "full")
  }
  
  total_big_merged_output$Condition <- sapply(str_split(total_big_merged_output$Sample, "[0-9]_"), `[`, 2)
  
  
  #Now create a new df but with avergae values for each condition.
  p5_pos <- unique(total_big_merged_output$p5)
  joined_df <- data.frame(p5=p5_pos)
  for ( sample in unique(total_big_merged_output$Sample)){
    subset_sample <- total_big_merged_output %>% filter(Sample==sample) %>% dplyr::select(p5, count)
    subset_sample <- subset_sample %>% dplyr::rename(!!paste0(sample):=count)
    joined_df <- join(joined_df,subset_sample, type="inner")
  }
  
  joined_df <- cbind( joined_df, met_liv = rowMeans( joined_df[ , grepl("met_liv", names(joined_df))] ),
                      sq = rowMeans( joined_df[ , grepl("sq", names(joined_df))]))
  
 
  mean_values <- c("met_liv","sq")
  
  new_df <- data.frame()
  for (i in mean_values){
    df <- joined_df %>% dplyr::select(p5,i)
    df$Condition <- rep(paste0(i), nrow(df))
    df <- df %>% dplyr::rename("counts"=2)
    df <- df[order(df$p5),]
    df$total <- sum(df$counts)
    df$fraction <- df$counts / df$total
    df$cummulative_counts <- cumsum(df$counts)
    df$cummulative_frac <- cumsum(df$fraction)
    new_df <- rbind(new_df,df)
  }
  #subset to keep only CDS coverage values.
  new_df <- new_df %>% filter(p5 >=cds_start, p5 <= cds_stop)
  
  #Now split df based on selenocystine location.
  five_prime_end <- new_df %>% dplyr::filter(p5 < sec_loc)
  three_prime_end <- new_df %>% dplyr::filter(p5 > sec_loc)
  # custom bins
  data <- cds_start:sec_loc
  # Number of intervals
  num_intervals <- 5
  # Calculate the width of each interval
  interval_width <- ceiling(length(data) / num_intervals)
  # Calculate the round breakpoints
  break_points <- seq(from = min(data), by = interval_width, length.out = num_intervals + 1)
  # Apply the cut function with the round breakpoints
  bins_five_prime <- cut(data, breaks = break_points, dig.lab = 5)
  level_bins_five_prime <- levels(bins_five_prime)
  # printing the subsets of dataframe
  five_prime_percentages <- data.frame()
  for(i in 1:length(level_bins_five_prime)) {  
    clean_string <- gsub("[^0-9]", " ", levels(bins_five_prime)[i])
    start <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,2]))
    print(start)
    end <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,3]))
    print(end)
    subset_dataframe <- five_prime_end %>% dplyr::filter(between(p5, start, end))
    #now that we have subseted by the bins of interest we will find the top fraction of both conditions and KO-WT
    ko_cond <- subset_dataframe %>% filter(Condition==grep("met_liv",mean_values, value = T))
    ko_max <- max(ko_cond$cummulative_frac)
    wt_cond <- subset_dataframe %>% filter(Condition==grep("sq",mean_values, value = T))
    wt_max <- max(wt_cond$cummulative_frac)
    value_lfc <- as.data.frame(ko_max-wt_max)
    value_lfc$percentage <- paste0(i)
    value_lfc$region <- paste0(start,":",end)
    value_lfc$footprint <- "five_prime"
    five_prime_percentages <- rbind(five_prime_percentages,value_lfc)
  }
  
  five_prime_percentages$percentage_real <- c("-100", "-80", "-60", "-40", "-20")
  
  #Now for three primer. Calculate transcript length-sec position to know the length.
  #Now split df based on selenocystine location.
  # custom bins
  data_threeprime <- 1:(cds_stop-sec_loc)
  data_threeprime <- data_threeprime + sec_loc
  # Calculate the width of each interval
  interval_width_three_prime <- ceiling(length(data_threeprime) / num_intervals)
  # Calculate the round breakpoints
  break_points_three_prime <- seq(from = min(data_threeprime), by = interval_width_three_prime, length.out = num_intervals + 1)
  # Apply the cut function with the round breakpoints
  bins_three_prime <- cut(data_threeprime, breaks = break_points_three_prime, dig.lab = 5)
  level_bins_three_prime <- levels(bins_three_prime)
  # printing the subsets of dataframe
  three_prime_percentages <- data.frame()
  for (i in 1:length(level_bins_three_prime)) {  
    clean_string <- gsub("[^0-9]", " ", levels(bins_three_prime)[i])
    start <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,2]))
    print(start)
    end <- round(as.numeric(str_split_fixed(clean_string, " ", 4)[,3]))
    print(end)
    subset_dataframe <- three_prime_end %>% dplyr::filter(between(p5, start, end))
    #now that we have subseted by the bins of interest we will find the top fraction of both conditions and KO-WT
    ko_cond <- subset_dataframe %>% filter(Condition==grep("met_liv",mean_values, value = T))
    ko_max <- max(ko_cond$cummulative_frac)
    wt_cond <- subset_dataframe %>% filter(Condition==grep("sq",mean_values, value = T))
    wt_max <- max(wt_cond$cummulative_frac)
    value_lfc_three <- as.data.frame(ko_max-wt_max)
    value_lfc_three$percentage <- paste0(i)
    value_lfc_three$region <- paste0(start,":",end)
    value_lfc_three$footprint <- "three_prime"
    three_prime_percentages <- rbind(three_prime_percentages,value_lfc_three)
  }
  three_prime_percentages[is.na(three_prime_percentages)] <- 0
  three_prime_percentages$percentage_real <- c("+20", "+40", "+60", "+80", "+100")
  
  
  all_together_one_seleno <- rbind(five_prime_percentages, three_prime_percentages)
  all_together_one_seleno$gene <- rep(paste0(gene_name), nrow(all_together_one_seleno))
  
  return(all_together_one_seleno)
  rm(list = ls())
  
  
  
}
```

#Use function in all selenoproteins and bind dfs.
```{r, echo=F, message=F, warning=F}
### GPX1
heatmap_df1 <- heatmap_values_seleno("ENST00000419783.3","GPX1",899,220, 76, 687)

### GPX4

heatmap_df3 <- heatmap_values_seleno("ENST00000354171.13","GPX4",851,267, 51, 644)

### TXNRD1
heatmap_df4 <- heatmap_values_seleno("ENST00000525566.6","TXNRD1",3860,1966, 25, 1724)

### TXNRD2
heatmap_df5 <- heatmap_values_seleno("ENST00000400521.7","TXNRD2",1941,1582, 16, 1590)

### MSRB1
heatmap_df6 <- heatmap_values_seleno("ENST00000361871.8","MSRB1",1277,326, 44, 394)

### SELENOH
heatmap_df7 <- heatmap_values_seleno("ENST00000534355.6","SELENOH",1192,235, 106, 474)
### SELENOI
heatmap_df8 <- heatmap_values_seleno("ENST00000260585.12","SELENOI",8066,1249, 91, 1284)

### SELENOK
heatmap_df9 <- heatmap_values_seleno("ENST00000495461.6","SELENOK",1497,345, 72, 356)

### SELENON
heatmap_df10 <- heatmap_values_seleno("ENST00000361547.7","SELENON",4314,417,39, 1709)
### SELENOO
heatmap_df11 <- heatmap_values_seleno("ENST00000380903.7","SELENOO",2283,2025, 27, 2036)

heatmap_df18 <- heatmap_values_seleno("ENST00000514985.6","SELENOP",2056,245, 71, 1214)


### SELENOS
heatmap_df12 <- heatmap_values_seleno("ENST00000526049.6","SELENOS",1218,360,69, 638)

### SELENOT
heatmap_df13 <- heatmap_values_seleno("ENST00000471696.6","SELENOT",3437,187,43, 630)

### SELENOW
heatmap_df14 <- heatmap_values_seleno("ENST00000601048.6","SELENOW",758,120,84, 347)

### SELENOF
heatmap_df15 <- heatmap_values_seleno("ENST00000331835.10","SELENOF",1542,301, 16, 513)

### SEPHS2
heatmap_df16 <- heatmap_values_seleno("ENST00000478753.5","SEPHS2",2244,325, 148, 1494)

### TXNRD3
heatmap_df17 <- heatmap_values_seleno("ENST00000524230.9","TXNRD3",2921,2058, 135, 2066)
```

```{r}
rm(all_heatmap_df)
df_list <-  mget(ls(pattern="heatmap_df"))
all_heatmap_df <- do.call(rbind, df_list)
write.csv(all_heatmap_df, "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Xenograft/p_site_coverage/heatmap_values/Xeno_cummulative_fraction_by_bin_prop.csv")
```

```{r}
subset_df <- all_heatmap_df %>% dplyr::select(`ko_max - wt_max`, percentage_real, gene)
names(subset_df) <- c("frac_change", "percentage_real", "gene")
top_genes <- subset_df %>% filter(percentage_real=="-20") %>% arrange(desc(frac_change))
order_genes <- top_genes$gene 

col_order <- c("-100", "-80", "-60", "-40", "-20", "+20",  "+40",  "+60",  "+80","+100")

subset_df_mat <- acast(subset_df, gene~percentage_real, value.var = "frac_change")
subset_df_mat <- subset_df_mat[order_genes,col_order]
```

# Heatmap options {.tabset}

## First annotation on the top

```{r, fig.height=8, fig.width=10}
heatmap1 <-Heatmap(subset_df_mat, name = "difference", cluster_columns =  F,cluster_rows = F ,column_split =rep(c("up", "down"), each=5),  column_title_gp = gpar(fill = c("lightcyan", "pink"), font = 1:1), column_gap = unit( 4, "mm"), column_names_rot = 0, column_labels = c("%100", "%80", "%60", "%40", "%20", "%20",  "%40",  "%60",  "%80","%100"))
heatmap1
```

## Second annotation in the bottom

```{r}
column_anno <- HeatmapAnnotation(Sec_pos = c("Up","Up","Up","Up","Up", "Down", "Down","Down","Down","Down" ),show_annotation_name = FALSE, col = list(Sec_pos = c("Up" = "lightcyan", "Down" = "pink")))
```

```{r, fig.height=8, fig.width=10}
heatmap2 <- Heatmap(subset_df_mat, name = "difference", cluster_columns =  F,cluster_rows = F ,column_split =rep(1:2, each=5),column_gap = unit( 4, "mm"), column_title = NULL,bottom_annotation = column_anno, column_labels = c("%100", "%80", "%60", "%40", "%20", "%20",  "%40",  "%60",  "%80","%100")) 
heatmap2
```


## {-}

```{r}
sessionInfo()
```
