---
title: "Script4 PDX readthrough based on 3' and 5' footprints"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output:
  html_document:
    df_print: paged
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',tidy=TRUE, cache=FALSE, echo = F)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```


```{r, echo=F, message=F}
#Load libraries
library(stringr)
library(plyr)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(grid)
library(tidyverse)
library(rstatix)
```

```{r, echo=F, message=F}
#For color change:
#https://www.nceas.ucsb.edu/sites/default/files/2020-04/colorPaletteCheatsheet.pdf
```

```{r, echo=F, message=F}
#From the output of Script1_ribosome_stalling.Rmd load the counts based on inframe psite coverage for each Selenoprotein.
#transcript name: Selenoprotein MANE v1.2 gencode annotation 
#gene name: gene symbol
#transcript length: transcript length 
#sec_loc: selenocysteine position at the transcript
#window_sec: bin length for each selenoprotein

```

The 5' and 3' prime footprints were determined based on inframe psite counts obtained from the CDF analysis. Using a fixed X bp window, we identified the regions corresponding to the 5' and 3' ends. Specifically, reads from the Sec codon position and below X bp were considered as 5' prime footprints, while reads from the Sec position and above X bp were considered as 3' prime footprints.

For each sample, we summed the reads within these respective regions. Then, for each condition, we calculated the average values of the footprints across the samples.


The comparisons we performed were:

3' prime / 5' prime: This ratio quantifies the relative abundance of 3' prime footprints compared to 5' prime footprints.
These comparisons provide insights into the distribution and relative abundance of footprints at the 5' and 3' ends.
```{r, echo=F, message=F}
footprint_df <- function(transcript_name,gene_name,transcript_length,sec_loc, window_sec){
   my_files <- list.files(paste0(cover_dir, "/",transcript_name), pattern=".cov")
  my_files2 <- str_extract(my_files, '.*(?=\\.cov)')
  #All the files in same format to read it correctly
  for (i in my_files2) {
    filename <- str_split(i, paste0(transcript_name, "_"))[[1]][2]
    wd <- paste0(cover_dir, "/", transcript_name, "/", i, ".cov")
    if (file.size(wd) > 0) {
      assign(filename, read.csv(wd, header=T, sep="\t"))
    } else {
      cat("File", wd, "is empty.\n")
    }
  }
  
  #Change the names of prob by each samples name.
  lst <-  mget(ls(pattern="cm2"))
  lst1 <- Map(function(x,y) {names(x)[1] <- y; x},  lst, paste0(gene_name))
  lst1 <- Map(function(x,y) {names(x)[2] <- y; x},  lst1, "p5")
  lst1 <- Map(function(x,y) {names(x)[3] <- y; x},  lst1, "count")
  
  
  list2env(lst1, envir=.GlobalEnv)
  
  total_rows <- sum(unlist(lapply(lst1, function(df) dim(df)[1])))
  #print(total_rows)
  total_col <- sum(unlist(lapply(lst1, function(df) dim(df)[2]))) #Since we have 18 samples *3
  #print(total_col)
  
  add_missing_rows <- function(df, MISSING, transcript_name, gene_name) {
    df <- rbind(df,setNames(data.frame(paste0(gene_name), MISSING, 0), c(paste(transcript_name), "p5", "count")))
    df[[""]] <- transcript_name
    df <- df[order(df$p5),]
    return(df)
  }
  
  lst1 <- lapply(lst1, function(df) {
    MISSING <- setdiff(seq(0, as.numeric(paste0(transcript_length)), 1), df$p5)
    df <- add_missing_rows(df, MISSING, gene_name, transcript_name)
    return(df)
  })
  
  total_big_merged_output <- data.frame()
  sample_names <- names(lst1)
  for (k in seq_along(lst1)) {
    i <- lst1[[k]]
    j <- sample_names[[k]]
    
    # Add a new column to the data frame based on sample name
    i$Sample <- rep(paste0(j), nrow(i))
    
    # Merge the data frame with total_big_merged_output
    total_big_merged_output <- plyr::join(total_big_merged_output, i, type = "full")
  }
  
  total_big_merged_output$Condition <- sapply(str_split(total_big_merged_output$Sample,  "_"), `[`, 2)
  
  total_big_merged_output <- total_big_merged_output %>% 
      dplyr::mutate(footprint = ifelse(p5 >= sec_loc, "3_prime", "5_prime"))

  
  #set up a window to calculate coverage through each selenoprotein.
  total_big_merged_output <- total_big_merged_output %>% filter(!(p5 > (sec_loc+window_sec) | p5 < (sec_loc-window_sec)))
  
  #now sum counts from sample sample and 3 and 5 prime.
  footprint_df <- total_big_merged_output %>%
    dplyr::group_by(Sample, footprint) %>%
    dplyr::summarize(sum_counts = sum(count)) %>%
    ungroup()
  
  # Step 1: Calculate total sum_counts for each sample
  total_counts <- footprint_df %>%
    group_by(Sample) %>%
    summarize(total_sum_counts = sum(sum_counts))
  
  
  footprint_df$Condition <-sapply(str_split(footprint_df$Sample,  "_"), `[`, 2)
  footprint_df$Gene <- rep(gene_name, nrow(footprint_df))
  return(footprint_df)
}


```

```{r, echo=F, message=F}
#RIBO location
cover_dir <- "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Xenograft/p_site_coverage"

footprint_df_results_list <- list()

footprint_df_results_list[["GPX1"]] <- footprint_df("ENST00000419783.3","GPX1",899,220, 200)

footprint_df_results_list[["GPX4"]] <- footprint_df("ENST00000354171.13","GPX4",851,267,  200)

footprint_df_results_list[["TXNRD1"]] <- footprint_df("ENST00000525566.6","TXNRD1",3860,1966,  6)

footprint_df_results_list[["TXNRD2"]] <- footprint_df("ENST00000400521.7","TXNRD2",1941,1582,  6)

footprint_df_results_list[["MSRB1"]] <- footprint_df("ENST00000361871.8","MSRB1",1277,326, 66)

footprint_df_results_list[["SELENOH"]] <- footprint_df("ENST00000534355.6","SELENOH",1192,235,  200)

footprint_df_results_list[["SELENOI"]] <- footprint_df("ENST00000260585.12","SELENOI",8066,1249,  33)

footprint_df_results_list[["SELENOK"]] <- footprint_df("ENST00000495461.6","SELENOK",1497,345, 9)

footprint_df_results_list[["SELENON"]] <- footprint_df("ENST00000361547.7","SELENON",4314,417,  200)

footprint_df_results_list[["SELENOO"]] <- footprint_df("ENST00000380903.7","SELENOO",2283,2025,  9)

footprint_df_results_list[["SELENOP"]] <- footprint_df("ENST00000514985.6","SELENOP",2056,245,  200)

footprint_df_results_list[["SELENOS"]] <- footprint_df("ENST00000526049.6","SELENOS",1218,360, 200)

footprint_df_results_list[["SELENOT"]] <- footprint_df("ENST00000471696.6","SELENOT",3437,187,  200)

footprint_df_results_list[["SELENOW"]] <- footprint_df("ENST00000601048.6","SELENOW",758,120,  100)

footprint_df_results_list[["SELENOF"]] <- footprint_df("ENST00000331835.10","SELENOF",1542,301, 200)

footprint_df_results_list[["SEPHS2"]]<- footprint_df("ENST00000478753.5","SEPHS2",2244,325, 200)

footprint_df_results_list[["TXNRD3"]] <- footprint_df("ENST00000524230.9","TXNRD3",2921,2058, 6)
```



```{r, echo=F, message=F}
complete_fottprint_df <- do.call(rbind,footprint_df_results_list)

```


## Frequency bar plots. {.tabset}

### Bar plot for RIBO fraction 3'/5' footprints for each gene.
```{r, echo=F, message=F}
write.table(complete_fottprint_df, "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Xenograft/footprints/Footprint_raw_sum_counts.csv", quote = F, sep = ",")

```

```{r, echo=F, message=F}

#Subset RIBO data.
fraction_RIBO_df <- data.frame()
for (i in unique(complete_fottprint_df$Gene)){
  freq_df <- complete_fottprint_df %>% filter(Gene==i)
  freq_df_2 <- freq_df %>%
    group_by(Sample) %>%
    filter(footprint %in% c("3_prime", "5_prime")) %>%
    pivot_wider(names_from = footprint, values_from = sum_counts) %>%
    mutate(Fraction = `3_prime` / `5_prime`)
  fraction_RIBO_df <- rbind(fraction_RIBO_df, freq_df_2)
}

fraction_RIBO_df[fraction_RIBO_df=="NaN"] <- 0
fraction_RIBO_df[fraction_RIBO_df=="Inf"] <- 0


fraction_RIBO_df3 <- fraction_RIBO_df %>% dplyr::arrange(Gene)
stat.test <- fraction_RIBO_df3 %>%
  group_by(Gene) %>%
  t_test(Fraction ~ Condition)
stat.test <- stat.test %>%
  add_xy_position(x = "Gene", dodge = 0.8)

write.csv(fraction_RIBO_df3, "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Xenograft/footprints/Footprint_fractions.csv")
```


```{r, echo=F, message=F}
# Create a bar plot with error bars (mean +/- sd)
plot <- ggbarplot(
  fraction_RIBO_df3, x = "Gene", y = "Fraction", add = "mean_sd", 
  color = "Condition", palette = c("salmon1", "slategray"),
  position = position_dodge(0.8)
  ) + stat_pvalue_manual(
  stat.test,  label = "p", tip.length = 0.01, label.size = 3,
  x = "Gene"
  ) + 
  labs(x = "Selenoproteins", 
       y = "3'RPF/5'RPF") + 
  theme(legend.title = element_text(size=12),
        legend.text = element_text(size=10)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(
    # Hide panel borders and remove grid lines
    panel.border = element_blank(),
    panel.background=element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # Change axis line
    axis.line = element_line(colour = "gray"))
print(plot)


```



## {-}


