---
title: "Script5 cell lines Translation Efficiency based on CDS counts"
output: html_document
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
---

We follow the analysis described in https://currentprotocols.onlinelibrary.wiley.com/doi/full/10.1002/cpmb.108 where they introduce an interaction term into the statistical model of DESeq2 to identify differential translation efficiency genes. When combining RNA-seq and Ribo-seq from two conditions, the interaction term can be used to model condition (untreated/treated) and sequencing methodology (Ribo-seq/RNA-seq). This allows the identification of significant differences between conditions that are discordant between sequencing methodologies. They design a generalized linear model with three components: the condition (c), the sequencing type (s), and an interaction term containing both (c:s). The result is a Î”TE fold change and an associated false discovery rate (FDR) for significant changes of this fold change, which quantify the extent of translational regulation between conditions.

```{r , echo=F, message=F, warning=F}
#Load libraraies
library(fgsea)
library(tximport)
library(openxlsx)
library(dplyr)
library(GenomicFeatures)
library(DESeq2)
library(stringr)
library(pheatmap)
library(plyr)
library(AnnotationDbi)
library(apeglm)
library(ggplot2)
library(EnhancedVolcano)
library(clusterProfiler)
library(org.Hs.eg.db)
library(msigdbr)
library(GGally)
library(tidyverse)
library(ggplotify)
library(ggpubr)
library(GGally)
library("ggVennDiagram")

```


```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, result='hold',fig.width=10,tidy=TRUE, cache=FALSE, echo = F)
knitr::opts_knit$set(progress=TRUE,verbose=TRUE)
```


```{r}
gtf.file = "/local/storage/maa7095/references/human/gencode_primary/MANE_ref/MANE_select_subset_GRCh38.v1.2.ensembl_genomic.gtf"
gtf.gr = rtracklayer::import(gtf.file) # creates a GRanges object
gtf.df = as.data.frame(gtf.gr)
genes = unique(gtf.df[ ,c("gene_id","gene_name", "transcript_id")])
```


```{r, echo=F, message=F, warning=F}
#Load the metadata.
TE_metadata <- read.xlsx("/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/TE_metadata_updated.xlsx")
ribo_metadata <- TE_metadata %>% filter(Data_type=="RIBO_seq")
rna_metadata <- TE_metadata %>% filter(Data_type=="RNA_seq")
```



```{r, echo=F, message=F, warning=F}
#Load RIBOseq data from ribowaltz inframe psite CDS counts.
RIBO_counts <- read.csv("/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/CDS_RIBO_raw_counts.csv")
#keep only trascript in MANE_select not clinical)
RIBO_counts <- RIBO_counts %>% filter(transcript %in% genes$transcript_id) %>% rename(transcript_id=transcript)

#rename transcript by gene symbol
genes <- genes %>% filter(transcript_id %in% RIBO_counts$transcript_id)
RIBO_counts <- join(RIBO_counts, genes, type="inner")
#Remove dupliacetd values
RIBO_counts = RIBO_counts[!duplicated(RIBO_counts$gene_name),]
rownames(RIBO_counts) <- RIBO_counts$gene_name
RIBO_counts <- RIBO_counts %>% dplyr::select(gene_name, contains("trimmed"))
RIBO_counts$gene_name <- NULL
names(RIBO_counts) <- sapply(strsplit(names(RIBO_counts), "_Aligned.toTranscriptome.out"), "[", 1)
names(RIBO_counts) <- sapply(strsplit(names(RIBO_counts), "human_mapped_"), "[", 2)
```

```{r, echo=F, message=F, warning=F}
#Load bulkRNAseq gene counts based on CDS from featurecounts.

RNA_counts <- read.table("/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/CDS_RNA_counts_not_strand.txt", header = T, sep = "\t")
RNA_counts <- RNA_counts %>% select(Geneid, contains("trimmed"))

RNA_counts <- RNA_counts %>% rename(gene_id=Geneid)
RNA_counts <- join(RNA_counts,genes)
RNA_counts = RNA_counts[!duplicated(RNA_counts$gene_name),]
RNA_counts <- RNA_counts %>% dplyr::select(gene_name, contains("trimmed"))
rownames(RNA_counts) <- RNA_counts$gene_name

RNA_counts$gene_name <- NULL
names(RNA_counts) <- sapply(strsplit(names(RNA_counts), "_Aligned.sortedByCoord.out.bam"), "[", 1)
names(RNA_counts) <- sapply(strsplit(names(RNA_counts), "human_mapped_"), "[", 2)
```

```{r, echo=F, message=F, warning=F}
TE_metadata$Data_type <- factor(TE_metadata$Data_type, levels = c("RNA_seq", "RIBO_seq"))
TE_metadata$Condition <- factor(TE_metadata$Condition, levels = c("WT_NO_SE","WT_SE","KO_NO_SE", "KO_SE"))
TE_metadata$Sample_type <- factor(TE_metadata$Sample_type, levels = c("wt_WT_NO_SE","wt_WT_SE", "ko1_KO_NO_SE", "ko1_KO_SE"))
```


```{r, echo=F, message=F, warning=F}
#Bind matrices and check for rows in order.
all(rownames(RIBO_counts)==rownames(RNA_counts))
RIBO_counts <- RIBO_counts[rownames(RNA_counts),]
all(rownames(RIBO_counts)==rownames(RNA_counts))
```


```{r, echo=F, message=F, warning=F}
TE_counts <- cbind(RIBO_counts,RNA_counts)
rownames(TE_metadata) <- TE_metadata$Sample_name
all(rownames(TE_metadata) == colnames(TE_counts))
TE_counts <- TE_counts[,rownames(TE_metadata)]
all(rownames(TE_metadata) == colnames(TE_counts))
```


```{r,  message=F, echo=F, warning=F}
#Start runnig DESEQ2.
TE_ddsMat = DESeqDataSetFromMatrix(
  countData=round(TE_counts),
  colData=TE_metadata,
  design=~ Sample_type+Data_type+Sample_type:Data_type)
```

```{r, echo=F, message=F}
keep <- rowSums(counts(TE_ddsMat) >= 10) >= 2
TE_ddsMat <- TE_ddsMat[keep,]
TE_ddsMat <- DESeq(TE_ddsMat)
TE_normalized_counts <- counts(TE_ddsMat,normalized=TRUE)
TE_raw_counts <- counts(TE_ddsMat,normalized=F)
# save raw counts and normalized counts
write.csv(TE_normalized_counts, "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/TE_normalized_counts.csv")
write.csv(TE_raw_counts, "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/TE_raw_counts.csv")

```


```{r, echo=F, message=F, warning=F}
#Subset dataset just for NO SE suppl. #The analysis will be done for absence or presence of Selenium separately.
subset_metadata <- TE_metadata %>% filter(Condition %in% c("WT_NO_SE","KO_NO_SE"))
subset_metadata$Condition <- factor(subset_metadata$Condition, levels = c("WT_NO_SE","KO_NO_SE"))
keep_NOSE_samples <- subset_metadata$Sample_name
```


```{r, echo=F, message=F, warning=F}
#Subset main object.
#Bind matrices and check for rows in order.
NOSE_TE_mat <- TE_counts[,keep_NOSE_samples]
all(rownames(subset_metadata) == colnames(NOSE_TE_mat))
```


```{r, echo=F, message=F, warning=F}
#Start runnig DESEQ2.
NOSE_TE_ddsMat = DESeqDataSetFromMatrix(
  countData=round(NOSE_TE_mat),
  colData=subset_metadata,
  design=~ Condition+Data_type+Condition:Data_type)
```

```{r, echo=F, message=F, warning=F}
keep <- rowSums(counts(NOSE_TE_ddsMat) >= 10) >= 2
NOSE_TE_ddsMat <- NOSE_TE_ddsMat[keep,]
NOSE_TE_ddsMat <- DESeq(NOSE_TE_ddsMat)
NOSE_TE_normalized_counts <- counts(NOSE_TE_ddsMat,normalized=TRUE)

write.csv(NOSE_TE_normalized_counts, "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/NOSE_TE_normalized_counts.csv")
```

## TE DE results.{.tabset}
First comparison in which three components: the condition (c), the sequencing type (s), and an interaction term containing both (c:s) will be take into account. Condition2.SeqTypeRibo.seq means Changes in Ribo-seq levels in Condition2 vs Condition1 accounting for changes in RNA-seq levels in Condition2 vs Condition1.

### PCA

```{r, echo=F, message=F, warning=F}
NOSE_TE_ddsMat_Coll <- rlog( NOSE_TE_ddsMat, blind=FALSE)
NOSE_TE_ddsMat_Coll_mat <- assay(NOSE_TE_ddsMat_Coll)
NOSE_TE_ddsMat_Coll_cor <- cor(NOSE_TE_ddsMat_Coll_mat)
plotPCA(NOSE_TE_ddsMat_Coll, intgroup=c("Condition", "Data_type")) 
```


```{r, echo=F, message=F, warning=F}
results_TE <- as.data.frame(results(NOSE_TE_ddsMat, contrast=list("ConditionKO_NO_SE.Data_typeRIBO_seq")))

#arrange by padj values
results_TE <- results_TE[order(results_TE[,6]),]
  
write.csv(results_TE, file="/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/NOSE_KO_TE_results/NO_SE_DESeq2_TE.csv", row.names = T, quote = F)
```


### Volcano plot KO RIBO vs WT RIBO and KO RNA vs WT RNA. 
```{r, echo=F, message=F, warning=F, warning=F}
EnhancedVolcano(results_TE,lab = rownames(results_TE), x = 'log2FoldChange',y = 'pvalue', title = "TE analysis", pCutoff = 1e-05, FCcutoff = 1)
```

### Heatmap showing expression of DE genes padj < 0.01 LFC 1
```{r, echo=F, message=F, warning=F}
#draw heatmap with the most disregulated genes
padj_genes <-na.omit(results_TE) %>% filter(padj<0.01 ,  log2FoldChange > 1 |  log2FoldChange < -1)
padj_genes2 <- rownames(padj_genes)
print(paste0(length(padj_genes2), " DE genes under padjusted < 0.01 and LFC 1 in TE data"))

#Get the transformed values and filter genes
subset_mat <- NOSE_TE_ddsMat_Coll_mat[padj_genes2,]
```

```{r, echo=F, message=F, warning=F}
pheatmap(subset_mat, scale="row", show_rownames = F, show_colnames = T, cluster_rows = T, legend = T, main =  "Heatmap of DE genes found in TE analysis", annotation_col = as.data.frame(colData(NOSE_TE_ddsMat))[, c("Data_type", "Condition")])
```


## {-}

# BulkRNA seq data. PCA of the data.

Second comparison KO_RNA vs WT_RNA.

```{r, echo=F, message=F}
#Create bulkRNAseq object.
rna_samples <-subset_metadata[which(subset_metadata$Data_type == "RNA_seq"),]$Sample_name 
RNA_ddsMat = DESeqDataSetFromMatrix(
  countData=round(NOSE_TE_mat[,rna_samples]),
  colData=subset_metadata[which(subset_metadata$Data_type == "RNA_seq"),],
  design=~Condition)
keep <- rowSums(counts(RNA_ddsMat) >= 10) >= 2
RNA_ddsMat <- RNA_ddsMat[keep,]
RNA_ddsMat <- DESeq(RNA_ddsMat)

RNA_ddsMat_Coll <- rlog( RNA_ddsMat, blind=FALSE)
RNA_ddsMat_mat <- assay(RNA_ddsMat_Coll)

plotPCA(RNA_ddsMat_Coll, intgroup="Condition") + coord_fixed(ratio = 5) 
```

```{r, echo=F, message=F}
#Run deseq2
res_rna = results(RNA_ddsMat, name="Condition_KO_NO_SE_vs_WT_NO_SE") 
res_rna = lfcShrink(RNA_ddsMat,res=res_rna, coef = "Condition_KO_NO_SE_vs_WT_NO_SE",  type="apeglm")
control_samples <- c("wt_noSe_RNA_1","wt_noSe_RNA_2", "wt_noSe_RNA_3")
print(paste0("printing control samples ", control_samples))
ko_samples <- c("ko1_noSe_RNA_1","ko1_noSe_RNA_2","ko1_noSe_RNA_3")
print(paste0("printing case samples ", ko_samples))
res_rna$RNA_WT_mean_values <- rowMeans( counts(RNA_ddsMat,normalized=TRUE)[,control_samples])
res_rna$RNA_KO1_mean_values <- rowMeans( counts(RNA_ddsMat,normalized=TRUE)[,ko_samples])
res_rna <- res_rna[order(res_rna[,6]),]
res_rna <- as.data.frame(res_rna)  
write.csv(res_rna, file="/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/NOSE_KO_TE_results/NO_SE_DESeq2_RNA.csv", row.names = T, quote = F)
```

## RNAseq KO1 vs WT DE results.{.tabset}

### Volcano plot KO1 vs WT. 
```{r, echo=F, message=F, warning=F}
EnhancedVolcano(res_rna,lab = rownames(res_rna), x = 'log2FoldChange',y = 'pvalue', title = "No Selenium Supplementation KO1 vs WT RNA results",  pCutoff = 1e-05, FCcutoff = 1)
```

### Heatmap showing expression of DE genes padj < 0.01 LFC 1
```{r, echo=F, message=F, warning=F}
#draw heatmap with the most disregulated genes
padj_genes <-na.omit(res_rna) %>% filter(padj<0.01 ,  log2FoldChange > 1 |  log2FoldChange < -1)

padj_genes2 <- rownames(padj_genes)
print(paste0(length(padj_genes2), " DE genes under p < 0.01 and LFC 1 in RNA_seq data"))
subset_mat <- RNA_ddsMat_mat[padj_genes2,]
needed_samples <- c(control_samples,ko_samples)
subset_mat2 <- subset_mat[,needed_samples]
pheatmap(subset_mat2, scale="row", show_rownames = F, show_colnames = T, cluster_rows = T, legend = T, main =  "Heatmap of DE genes found in bulkRNAseq analysis", annotation_col = as.data.frame(colData(RNA_ddsMat))[, c("Data_type", "Condition")])
```

## {-}


Third comparison KO_RIBO vs WT_RIBO.


```{r, echo=F, message=F}
# RIBO seq data. PCA of the data.
#Create RIBOseq object.
RIBO_samples <-subset_metadata[which(subset_metadata$Data_type == "RIBO_seq"),]$Sample_name 
RIBO_ddsMat = DESeqDataSetFromMatrix(
  countData=round(NOSE_TE_mat[,RIBO_samples]),
  colData=subset_metadata[which(subset_metadata$Data_type == "RIBO_seq"),],
  design=~Condition)
keep <- rowSums(counts(RIBO_ddsMat) >= 10) >= 2
RIBO_ddsMat <- RIBO_ddsMat[keep,]
RIBO_ddsMat <- DESeq(RIBO_ddsMat)

RIBO_ddsMat_Coll <- rlog( RIBO_ddsMat, blind=FALSE)
RIBO_ddsMat_mat <- assay(RIBO_ddsMat_Coll)

plotPCA(RIBO_ddsMat_Coll, intgroup="Condition") + coord_fixed(ratio = 5) 
```

```{r, echo=F, message=F}
#Run deseq2
res_RIBO = results(RIBO_ddsMat, name="Condition_KO_NO_SE_vs_WT_NO_SE") 
res_RIBO = lfcShrink(RIBO_ddsMat,res=res_RIBO, coef = "Condition_KO_NO_SE_vs_WT_NO_SE",  type="apeglm")
control_samples <- c("wt_no_se_RIBO_1","wt_no_se_RIBO_2", "wt_no_se_RIBO_3")
print(paste0("printing control samples ", control_samples))
ko_samples <- c("ko1_no_se_RIBO_1","ko1_no_se_RIBO_2","ko1_no_se_RIBO_3")
print(paste0("printing case samples ", ko_samples))
res_RIBO$RIBO_WT_mean_values <- rowMeans( counts(RIBO_ddsMat,normalized=TRUE)[,control_samples])
res_RIBO$RIBO_KO1_mean_values <- rowMeans( counts(RIBO_ddsMat,normalized=TRUE)[,ko_samples])
res_RIBO <- res_RIBO[order(res_RIBO[,6]),]
res_RIBO <- as.data.frame(res_RIBO)  
write.csv(res_RIBO, file="/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/NOSE_KO_TE_results/NO_SE_DESeq2_RIBO.csv", row.names = T, quote = F)
```

## RIBOseq KO1 vs WT DE results.{.tabset}

### Volcano plot KO1 vs WT. 
```{r, echo=F, message=F, warning=F}
EnhancedVolcano(res_RIBO,lab = rownames(res_RIBO), x = 'log2FoldChange',y = 'pvalue', title = "No Sel Supplementation KO vs WT RIBO results",  pCutoff = 1e-05, FCcutoff = 1)
```

### Heatmap showing expression of DE genes padj < 0.01 LFC 1
```{r, echo=F, message=F, warning=F}
#draw heatmap with the most disregulated genes
padj_genes <-na.omit(res_RIBO) %>% filter(padj<0.01 ,  log2FoldChange > 1 |  log2FoldChange < -1)

padj_genes2 <- rownames(padj_genes)
print(paste0(length(padj_genes2), " DE genes under p < 0.01 and LFC 1 in RIBO_seq data"))

#Get the transformed values and filter genes
subset_mat <- RIBO_ddsMat_mat[padj_genes2,]

#Select the samples for each Condition
needed_samples <- c(control_samples,ko_samples)
subset_mat2 <- subset_mat[,needed_samples]
pheatmap(subset_mat2, scale="row", show_rownames = F, show_colnames = T, cluster_rows = T, legend = T, main =  "Heatmap of DE genes found in RIBOseq analysis", annotation_col =as.data.frame(colData(RIBO_ddsMat))[, c("Data_type", "Condition")])
```

## {-}



# Summary of TE for each selenoproteins {.tabset}

For each sample do log2(RIBO)-log(RNA) and plot the values. Across these values add the LFC of KO vs WT obtained by DESeq2. Therefore: 

(log2(RIBO_ko)-log(RNA_ko)) - (log2(RIBO_wt)-log(RNA_wt))

Sample values should be in accordance with the values obtained by DESeq2.


## NO selenium Supplementation.

```{r}
TE_NO_SE <- NOSE_TE_normalized_counts
rownames(TE_NO_SE) <- TE_NO_SE$X
TE_NO_SE$X<- NULL
colnames(TE_NO_SE) <- str_replace_all(colnames(TE_NO_SE), "noSe", "no_se")
new_names <- c("RIBO_seq_ko1_no_se_1","RIBO_seq_ko1_no_se_2","RIBO_seq_ko1_no_se_3", "RIBO_seq_wt_no_se_1","RIBO_seq_wt_no_se_2", "RIBO_seq_wt_no_se_3", "RNA_seq_ko1_no_se_1", "RNA_seq_ko1_no_se_2", "RNA_seq_ko1_no_se_3", "RNA_seq_wt_no_se_1", "RNA_seq_wt_no_se_2", "RNA_seq_wt_no_se_3")
colnames(TE_NO_SE) <- new_names
wt_TE_NO_SE <- TE_NO_SE %>% dplyr::select(contains("wt"))
ko1_TE_NO_SE <- TE_NO_SE %>% dplyr::select(contains("ko1"))
ko_cell_list <- c("no_se_1", "no_se_2", "no_se_3")
seleno_list <- c("GPX1","GPX2", "GPX3", "GPX4", "GPX6", "TXNRD1", "TXNRD2","DIO1", "DIO2", "DIO3", "MSRB1", "SELENOH", "SELENOI", "SELENOK", "SELENOM","SELENON", "SELENOO","SELENOP","SELENOS","SELENOT","SELENOV","SELENOW","SELENOF","SEPHS2","TXNRD3")

```


```{r}
TE_values_wt <- data.frame()
for (gene in seleno_list){
  test <- wt_TE_NO_SE[gene,]
  sample_data <- data.frame(gene=paste0(gene))
  for (ko_cell in ko_cell_list){
    sub_test <- test %>% dplyr::select(contains(ko_cell))
    res <- as.data.frame(log2(sub_test[,grep("RIBO", colnames(sub_test), value = T)])- log2(sub_test[,grep("RNA", colnames(sub_test), value = T)]))
    res <- res %>% dplyr::rename(!!paste0("WT_",ko_cell):=1)
    sample_data <- cbind(sample_data, res)
  }
  TE_values_wt <- rbind(TE_values_wt, sample_data)
}
TE_values_wt[TE_values_wt=="-Inf"] <- 0
```

```{r}
ko_cell_list <- c("no_se_1", "no_se_2", "no_se_3")
TE_values_ko <- data.frame()
for (gene in seleno_list){
  test <- ko1_TE_NO_SE[gene,]
  sample_data <- data.frame(gene=paste0(gene))
  for (ko_cell in ko_cell_list){
    sub_test <- test %>% dplyr::select(contains(ko_cell))
    res <- as.data.frame(log2(sub_test[,grep("RIBO", colnames(sub_test), value = T)])- log2(sub_test[,grep("RNA", colnames(sub_test), value = T)]))
    res <- res %>% dplyr::rename(!!paste0("KO_",ko_cell):=1)
    sample_data <- cbind(sample_data, res)
  }
  TE_values_ko <- rbind(TE_values_ko, sample_data)
}
TE_values_ko[TE_values_ko=="-Inf"] <- 0
```

```{r}
TE_table_by_sample <- join(TE_values_ko, TE_values_wt, type="inner")
TE_table_by_sample <- na.omit(TE_table_by_sample)

```

```{r}
#Check if individual computed fold changes values match DESeq2 results
mod_TE_table_by_sample <- TE_table_by_sample
rownames(mod_TE_table_by_sample) <- mod_TE_table_by_sample$gene
mod_TE_table_by_sample$gene <- NULL
rowMeans(mod_TE_table_by_sample["GPX4" ,(grep("KO", colnames(mod_TE_table_by_sample), value=T))]) - rowMeans(mod_TE_table_by_sample["GPX4" ,(grep("WT", colnames(mod_TE_table_by_sample), value=T))]) #same for other selenoproteins
```


```{r}
TE_NO_SE <- results_TE
TE_NO_SE$class_No_Sup <- rep("KO/Control-NoSe", nrow(TE_NO_SE))
subset_TE_NO_SE <- TE_NO_SE %>% dplyr::rename(TE_LFC =log2FoldChange, gene=X) %>% dplyr::filter(gene %in% seleno_list) %>% dplyr::select(gene, TE_LFC)
TE_table_by_sample <- join(TE_table_by_sample,subset_TE_NO_SE, type="inner") 
TE_table_by_sample <- TE_table_by_sample %>% arrange(TE_LFC)
rownames(TE_table_by_sample) <- TE_table_by_sample$gene
annotation_row = data.frame(TE_table_by_sample$TE_LFC)
names(annotation_row) <- "TE_LFC"
TE_table_by_sample$gene <- NULL

write.csv(TE_table_by_sample, "/SSD/maa7095/RIBOseq/Shino_new_analysis/inframe_downtream_analysis/Cell_lines/TE_analysis/NOSE_KO_TE_results/table_for_heatmaps/NO_Se_TE_table.csv")
```


```{r, fig.height=10, fig.width=10}
TE_table_by_sample <- as.matrix(TE_table_by_sample[,1:6])

annotation_col = data.frame(
    Condition=c("KO", "KO", "KO", "WT", "WT", "WT") 
)

condition <- data.frame(Condition=c("KO", "KO", "KO", "WT", "WT", "WT"))
order_samples <- colnames(TE_table_by_sample)
paletteLength <- 50
myColor <- colorRampPalette(c("navy", "white", "red"))(paletteLength)

myBreaks <- c(seq(min(TE_table_by_sample), 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(TE_table_by_sample)/paletteLength, max(TE_table_by_sample), length.out=floor(paletteLength/2)))

annotation_colors = list(
  Condition = c(KO="salmon1", WT="slategray"),
  TE_LCF = c("white", "firebrick")
  )

ann_colors = list(
    Time = c("white", "firebrick"),
    CellType = c(CT1 = "#1B9E77", CT2 = "#D95F02"),
    GeneClass = c(Path1 = "#7570B3", Path2 = "#E7298A", Path3 = "#66A61E")
)

pheatmap(TE_table_by_sample, cluster_col = F,  main =  "Translation Efficiency in No Selenium Supplementation", breaks = myBreaks, color = myColor, annotation_col = as.data.frame(condition), annotation_row =annotation_row, annotation_colors = annotation_colors, cellwidth = 25, cellheight = 25,  border_color=NA, treeheight_row=0,treeheight_col=0,cluster_rows = F , name="TE")
```



## {-}

# Session info.
```{r, echo=F, message=F}
sessionInfo()
```

